--- net/minecraft/src/MouseHelper.java
+++ net/minecraft/src/MouseHelper.java
@@ -5,6 +5,7 @@
 import org.lwjgl.LWJGLException;
 import org.lwjgl.input.Cursor;
 import org.lwjgl.input.Mouse;
+import org.lwjgl.opengl.Display;
 
 public class MouseHelper {
 	private Component field_1117_c;
@@ -35,7 +36,12 @@
 	}
 
 	public void ungrabMouseCursor() {
-		Mouse.setCursorPosition(this.field_1117_c.getWidth() / 2, this.field_1117_c.getHeight() / 2);
+		if(Display.isFullscreen()) {
+			Mouse.setCursorPosition(Display.getDisplayMode().getWidth() / 2, Display.getDisplayMode().getHeight() / 2);
+		} else {
+			Mouse.setCursorPosition(this.field_1117_c.getWidth() / 2, this.field_1117_c.getHeight() / 2);
+		}
+		
 		Mouse.setGrabbed(false);
 	}
 
