--- net/minecraft/src/ItemFood.java
+++ net/minecraft/src/ItemFood.java
@@ -1,8 +1,11 @@
 package net.minecraft.src;
 
+import java.util.Random;
+
 public class ItemFood extends Item {
 	private int healAmount;
 	private boolean isWolfsFavoriteMeat;
+	private static final Random random = new Random();
 
 	public ItemFood(int var1, int var2, boolean var3) {
 		super(var1);
@@ -14,6 +17,13 @@
 	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
 		--var1.stackSize;
 		var3.heal(this.healAmount);
+
+		var2.playSoundAtEntity(var3, "random.eat", 1.0F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F);
+
+		if(this.healAmount + var3.health >= 20) {
+			var2.playSoundAtEntity(var3, "random.burp", 0.5F, (random.nextFloat() - random.nextFloat()) * 0.2F + 1.0F);
+		}
+
 		return var1;
 	}
 
