--- net/minecraft/src/RenderBlocks.java
+++ net/minecraft/src/RenderBlocks.java
@@ -246,19 +246,19 @@
 		}
 
 		var6.setColorOpaque_F(var7, var7, var7);
-		double var8 = (double)0.4F;
-		double var10 = 0.5D - var8;
-		double var12 = (double)0.2F;
+		float var8 = (float)Math.PI * 0.125F;
+		double var10 = 0.5D;
+		double var12 = 0.21875D;
 		if(var5 == 1) {
-			this.renderTorchAtAngle(var1, (double)var2 - var10, (double)var3 + var12, (double)var4, -var8, 0.0D);
+			this.renderTorchAtAngle(var1, (double)var2 - var10, (double)var3 + var12, (double)var4, 0.0F, -var8);
 		} else if(var5 == 2) {
-			this.renderTorchAtAngle(var1, (double)var2 + var10, (double)var3 + var12, (double)var4, var8, 0.0D);
+			this.renderTorchAtAngle(var1, (double)var2 + var10, (double)var3 + var12, (double)var4, 0.0F, var8);
 		} else if(var5 == 3) {
-			this.renderTorchAtAngle(var1, (double)var2, (double)var3 + var12, (double)var4 - var10, 0.0D, -var8);
+			this.renderTorchAtAngle(var1, (double)var2, (double)var3 + var12, (double)var4 - var10, -var8, 0.0F);
 		} else if(var5 == 4) {
-			this.renderTorchAtAngle(var1, (double)var2, (double)var3 + var12, (double)var4 + var10, 0.0D, var8);
+			this.renderTorchAtAngle(var1, (double)var2, (double)var3 + var12, (double)var4 + var10, var8, 0.0F);
 		} else {
-			this.renderTorchAtAngle(var1, (double)var2, (double)var3, (double)var4, 0.0D, 0.0D);
+			this.renderTorchAtAngle(var1, (double)var2, (double)var3, (double)var4, 0.0F, 0.0F);
 		}
 
 		return true;
@@ -299,8 +299,8 @@
 			var12 = BlockRedstoneRepeater.field_22024_a[var7];
 		}
 
-		this.renderTorchAtAngle(var1, (double)var2 + var12, (double)var3 + var10, (double)var4 + var14, 0.0D, 0.0D);
-		this.renderTorchAtAngle(var1, (double)var2 + var16, (double)var3 + var10, (double)var4 + var18, 0.0D, 0.0D);
+		this.renderTorchAtAngle(var1, (double)var2 + var12, (double)var3 + var10, (double)var4 + var14, 0.0F, 0.0F);
+		this.renderTorchAtAngle(var1, (double)var2 + var16, (double)var3 + var10, (double)var4 + var18, 0.0F, 0.0F);
 		int var20 = var1.getBlockTextureFromSide(1);
 		int var21 = (var20 & 15) << 4;
 		int var22 = var20 & 240;
@@ -1332,51 +1332,120 @@
 		return true;
 	}
 
-	public void renderTorchAtAngle(Block var1, double var2, double var4, double var6, double var8, double var10) {
-		Tessellator var12 = Tessellator.instance;
-		int var13 = var1.getBlockTextureFromSide(0);
+	public void renderTorchAtAngle(Block var1, double var2, double var4, double var6, float var8, float var10) {
+		Tessellator var68 = Tessellator.instance;
+		int var11 = var1.getBlockTextureFromSide(0);
 		if(this.overrideBlockTexture >= 0) {
-			var13 = this.overrideBlockTexture;
-		}
-
-		int var14 = (var13 & 15) << 4;
-		int var15 = var13 & 240;
-		float var16 = (float)var14 / 256.0F;
-		float var17 = ((float)var14 + 15.99F) / 256.0F;
-		float var18 = (float)var15 / 256.0F;
-		float var19 = ((float)var15 + 15.99F) / 256.0F;
-		double var20 = (double)var16 + 1.75D / 64.0D;
-		double var22 = (double)var18 + 6.0D / 256.0D;
-		double var24 = (double)var16 + 9.0D / 256.0D;
-		double var26 = (double)var18 + 1.0D / 32.0D;
-		var2 += 0.5D;
-		var6 += 0.5D;
-		double var28 = var2 - 0.5D;
-		double var30 = var2 + 0.5D;
-		double var32 = var6 - 0.5D;
-		double var34 = var6 + 0.5D;
-		double var36 = 1.0D / 16.0D;
-		double var38 = 0.625D;
-		var12.addVertexWithUV(var2 + var8 * (1.0D - var38) - var36, var4 + var38, var6 + var10 * (1.0D - var38) - var36, var20, var22);
-		var12.addVertexWithUV(var2 + var8 * (1.0D - var38) - var36, var4 + var38, var6 + var10 * (1.0D - var38) + var36, var20, var26);
-		var12.addVertexWithUV(var2 + var8 * (1.0D - var38) + var36, var4 + var38, var6 + var10 * (1.0D - var38) + var36, var24, var26);
-		var12.addVertexWithUV(var2 + var8 * (1.0D - var38) + var36, var4 + var38, var6 + var10 * (1.0D - var38) - var36, var24, var22);
-		var12.addVertexWithUV(var2 - var36, var4 + 1.0D, var32, (double)var16, (double)var18);
-		var12.addVertexWithUV(var2 - var36 + var8, var4 + 0.0D, var32 + var10, (double)var16, (double)var19);
-		var12.addVertexWithUV(var2 - var36 + var8, var4 + 0.0D, var34 + var10, (double)var17, (double)var19);
-		var12.addVertexWithUV(var2 - var36, var4 + 1.0D, var34, (double)var17, (double)var18);
-		var12.addVertexWithUV(var2 + var36, var4 + 1.0D, var34, (double)var16, (double)var18);
-		var12.addVertexWithUV(var2 + var8 + var36, var4 + 0.0D, var34 + var10, (double)var16, (double)var19);
-		var12.addVertexWithUV(var2 + var8 + var36, var4 + 0.0D, var32 + var10, (double)var17, (double)var19);
-		var12.addVertexWithUV(var2 + var36, var4 + 1.0D, var32, (double)var17, (double)var18);
-		var12.addVertexWithUV(var28, var4 + 1.0D, var6 + var36, (double)var16, (double)var18);
-		var12.addVertexWithUV(var28 + var8, var4 + 0.0D, var6 + var36 + var10, (double)var16, (double)var19);
-		var12.addVertexWithUV(var30 + var8, var4 + 0.0D, var6 + var36 + var10, (double)var17, (double)var19);
-		var12.addVertexWithUV(var30, var4 + 1.0D, var6 + var36, (double)var17, (double)var18);
-		var12.addVertexWithUV(var30, var4 + 1.0D, var6 - var36, (double)var16, (double)var18);
-		var12.addVertexWithUV(var30 + var8, var4 + 0.0D, var6 - var36 + var10, (double)var16, (double)var19);
-		var12.addVertexWithUV(var28 + var8, var4 + 0.0D, var6 - var36 + var10, (double)var17, (double)var19);
-		var12.addVertexWithUV(var28, var4 + 1.0D, var6 - var36, (double)var17, (double)var18);
+			var11 = this.overrideBlockTexture;
+		}
+
+		int var12 = (var11 & 15) << 4;
+		int var13 = var11 & 240;
+		float var14 = (float)var12 / 256.0F;
+		float var15 = ((float)var12 + 15.99F) / 256.0F;
+		float var16 = (float)var13 / 256.0F;
+		float var17 = ((float)var13 + 15.99F) / 256.0F;
+		Vec3D[] var18 = new Vec3D[24];
+		float var19 = 1.0F / 16.0F;
+		float var20 = 1.0F / 16.0F;
+		float var21 = 10.0F / 16.0F;
+
+		int var22;
+		for(var22 = 0; var22 < 3; ++var22) {
+			if(var22 == 1) {
+				var19 = 0.5F;
+				var20 = 1.0F / 16.0F;
+				var21 = 1.0F;
+			} else if(var22 == 2) {
+				var19 = 1.0F / 16.0F;
+				var20 = 0.5F;
+			}
+
+			var18[0 + var22 * 8] = Vec3D.createVector((double)(-var19), 0.0D, (double)(-var20));
+			var18[1 + var22 * 8] = Vec3D.createVector((double)var19, 0.0D, (double)(-var20));
+			var18[2 + var22 * 8] = Vec3D.createVector((double)var19, 0.0D, (double)var20);
+			var18[3 + var22 * 8] = Vec3D.createVector((double)(-var19), 0.0D, (double)var20);
+			var18[4 + var22 * 8] = Vec3D.createVector((double)(-var19), (double)var21, (double)(-var20));
+			var18[5 + var22 * 8] = Vec3D.createVector((double)var19, (double)var21, (double)(-var20));
+			var18[6 + var22 * 8] = Vec3D.createVector((double)var19, (double)var21, (double)var20);
+			var18[7 + var22 * 8] = Vec3D.createVector((double)(-var19), (double)var21, (double)var20);
+		}
+
+		for(var22 = 0; var22 < 24; ++var22) {
+			var18[var22].rotateAroundX(var8);
+			float var23 = MathHelper.cos(var10);
+			float var24 = MathHelper.sin(var10);
+			double var25 = var18[var22].xCoord * (double)var23 - var18[var22].yCoord * (double)var24;
+			double var27 = var18[var22].yCoord * (double)var23 + var18[var22].xCoord * (double)var24;
+			double var29 = var18[var22].zCoord;
+			var18[var22].xCoord = var25;
+			var18[var22].yCoord = var27;
+			var18[var22].zCoord = var29;
+			var18[var22].xCoord += var2 + 0.5D;
+			var18[var22].yCoord += var4;
+			var18[var22].zCoord += var6 + 0.5D;
+		}
+
+		Vec3D var31 = null;
+		Vec3D var32 = null;
+		Vec3D var33 = null;
+		Vec3D var34 = null;
+
+		for(int var26 = 0; var26 < 6; ++var26) {
+			if(var26 == 0) {
+				var14 = (float)(var12 + 7) / 256.0F;
+				var15 = ((float)(var12 + 9) - 0.01F) / 256.0F;
+				var16 = ((float)(var13 + 15) - 0.01F) / 256.0F;
+				var17 = (float)(var13 + 13) / 256.0F;
+			} else if(var26 == 1) {
+				var14 = (float)(var12 + 7) / 256.0F;
+				var15 = ((float)(var12 + 9) - 0.01F) / 256.0F;
+				var16 = (float)(var13 + 6) / 256.0F;
+				var17 = ((float)(var13 + 8) - 0.01F) / 256.0F;
+			} else if(var26 == 2) {
+				var14 = (float)var12 / 256.0F;
+				var15 = ((float)(var12 + 16) - 0.01F) / 256.0F;
+				var16 = (float)var13 / 256.0F;
+				var17 = ((float)(var13 + 16) - 0.01F) / 256.0F;
+			}
+
+			if(var26 == 0) {
+				var31 = var18[0];
+				var32 = var18[1];
+				var33 = var18[2];
+				var34 = var18[3];
+			} else if(var26 == 1) {
+				var31 = var18[7];
+				var32 = var18[6];
+				var33 = var18[5];
+				var34 = var18[4];
+			} else if(var26 == 2) {
+				var31 = var18[9];
+				var32 = var18[8];
+				var33 = var18[12];
+				var34 = var18[13];
+			} else if(var26 == 3) {
+				var31 = var18[18];
+				var32 = var18[17];
+				var33 = var18[21];
+				var34 = var18[22];
+			} else if(var26 == 4) {
+				var31 = var18[11];
+				var32 = var18[10];
+				var33 = var18[14];
+				var34 = var18[15];
+			} else if(var26 == 5) {
+				var31 = var18[16];
+				var32 = var18[19];
+				var33 = var18[23];
+				var34 = var18[20];
+			}
+
+			var68.addVertexWithUV(var31.xCoord, var31.yCoord, var31.zCoord, (double)var14, (double)var17);
+			var68.addVertexWithUV(var32.xCoord, var32.yCoord, var32.zCoord, (double)var15, (double)var17);
+			var68.addVertexWithUV(var33.xCoord, var33.yCoord, var33.zCoord, (double)var15, (double)var16);
+			var68.addVertexWithUV(var34.xCoord, var34.yCoord, var34.zCoord, (double)var14, (double)var16);
+		}
 	}
 
 	public void renderCrossedSquares(Block var1, int var2, double var3, double var5, double var7) {
@@ -3242,7 +3311,7 @@
 			} else if(var5 == 2) {
 				var4.startDrawingQuads();
 				var4.setNormal(0.0F, -1.0F, 0.0F);
-				this.renderTorchAtAngle(var1, -0.5D, -0.5D, -0.5D, 0.0D, 0.0D);
+				this.renderTorchAtAngle(var1, -0.5D, -0.5D, -0.5D, 0.0F, 0.0F);
 				var4.draw();
 			} else {
 				int var9;
