--- net/minecraft/src/ItemFood.java
+++ net/minecraft/src/ItemFood.java
@@ -14,6 +14,11 @@
 	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
 		--var1.stackSize;
 		var3.heal(this.healAmount);
+		var2.playSoundAtEntity(var3, "random.eat", 0.5F + 0.5F * (float)var3.rand.nextInt(2), (var3.rand.nextFloat() - var3.rand.nextFloat()) * 0.2F + 1.0F);
+		if(this.healAmount + var3.health >= 20) {
+			var2.playSoundAtEntity(var3, "random.burp", 0.5F, var2.rand.nextFloat() * 0.1F + 0.9F);
+		}
+		
 		return var1;
 	}
 
--- net/minecraft/src/EntityPlayerSP.java
+++ net/minecraft/src/EntityPlayerSP.java
@@ -106,6 +106,10 @@
 	}
 
 	public void closeScreen() {
+		if(this.mc.currentScreen instanceof GuiChest) {
+			this.worldObj.playSoundAtEntity(this, "random.door_close", 1.0F, this.worldObj.rand.nextFloat() * 0.1F + 0.9F);
+		}
+
 		super.closeScreen();
 		this.mc.displayGuiScreen((GuiScreen)null);
 	}
@@ -116,6 +120,7 @@
 
 	public void displayGUIChest(IInventory var1) {
 		this.mc.displayGuiScreen(new GuiChest(this.inventory, var1));
+		this.worldObj.playSoundAtEntity(this, "random.door_open", 1.0F, this.worldObj.rand.nextFloat() * 0.1F + 0.9F);
 	}
 
 	public void displayWorkbenchGUI(int var1, int var2, int var3) {
--- net/minecraft/src/ItemStack.java
+++ net/minecraft/src/ItemStack.java
@@ -126,7 +126,12 @@
 			this.itemDamage += var1;
 			if(this.itemDamage > this.getMaxDamage()) {
 				if(var2 instanceof EntityPlayer) {
+					EntityPlayer var3 = (EntityPlayer) var2;
 					((EntityPlayer)var2).addStat(StatList.field_25170_B[this.itemID], 1);
+
+					if(this.getItem() instanceof ItemTool || this.getItem() instanceof ItemArmor) {
+						var3.worldObj.playSoundAtEntity(var3, "random.break", 0.8F, 0.8F + var3.worldObj.rand.nextFloat() * 0.4F);
+					}
 				}
 
 				--this.stackSize;
--- net/minecraft/src/EntitySheep.java
+++ net/minecraft/src/EntitySheep.java
@@ -47,6 +47,7 @@
 			}
 
 			var2.damageItem(1, var1);
+			this.worldObj.playSoundAtEntity(this, "mob.sheep.shear", 1.0F, 1.0F);
 		}
 
 		return false;
--- net/minecraft/src/EntityLiving.java
+++ net/minecraft/src/EntityLiving.java
@@ -441,6 +441,12 @@
 		super.fall(var1);
 		int var2 = (int)Math.ceil((double)(var1 - 3.0F));
 		if(var2 > 0) {
+			if(var2 > 4) {
+				this.worldObj.playSoundAtEntity(this, "damage.fallbig", 1.0F, 1.0F);
+			} else {
+				this.worldObj.playSoundAtEntity(this, "damage.fallsmall", 1.0F, 1.0F);
+			}
+			
 			this.attackEntityFrom((Entity)null, var2);
 			int var3 = this.worldObj.getBlockId(MathHelper.floor_double(this.posX), MathHelper.floor_double(this.posY - (double)0.2F - (double)this.yOffset), MathHelper.floor_double(this.posZ));
 			if(var3 > 0) {
