--- /dev/null
+++ net/minecraft/src/EntityLargeExplodeFX.java
@@ -1,0 +1,66 @@
+package net.minecraft.src;
+
+import org.lwjgl.opengl.GL11;
+
+public class EntityLargeExplodeFX extends EntityFX {
+	private int field_35130_a = 0;
+	private int field_35129_ay = 0;
+	private RenderEngine field_35128_az;
+	private float field_35131_aA;
+
+	public EntityLargeExplodeFX(RenderEngine var1, World var2, double var3, double var5, double var7, double var9, double var11, double var13) {
+		super(var2, var3, var5, var7, 0.0D, 0.0D, 0.0D);
+		this.field_35128_az = var1;
+		this.field_35129_ay = 6 + this.rand.nextInt(4);
+		this.particleRed = this.particleGreen = this.particleBlue = this.rand.nextFloat() * 0.6F + 0.4F;
+		this.field_35131_aA = 1.0F - (float)var9 * 0.5F;
+	}
+
+	public void renderParticle(Tessellator var1, float var2, float var3, float var4, float var5, float var6, float var7) {
+		int var8 = (int)(((float)this.field_35130_a + var2) * 15.0F / (float)this.field_35129_ay);
+		if(var8 <= 15) {
+			this.field_35128_az.bindTexture(this.field_35128_az.getTexture("/misc/explosion.png"));
+			float var9 = (float)(var8 % 4) / 4.0F;
+			float var10 = var9 + 0.999F / 4.0F;
+			float var11 = (float)(var8 / 4) / 4.0F;
+			float var12 = var11 + 0.999F / 4.0F;
+			float var13 = 2.0F * this.field_35131_aA;
+			float var14 = (float)(this.prevPosX + (this.posX - this.prevPosX) * (double)var2 - interpPosX);
+			float var15 = (float)(this.prevPosY + (this.posY - this.prevPosY) * (double)var2 - interpPosY);
+			float var16 = (float)(this.prevPosZ + (this.posZ - this.prevPosZ) * (double)var2 - interpPosZ);
+			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+			GL11.glDisable(GL11.GL_LIGHTING);
+			RenderHelper.disableStandardItemLighting();
+			var1.startDrawingQuads();
+			var1.setColorRGBA_F(this.particleRed, this.particleGreen, this.particleBlue, 1.0F);
+			var1.setNormal(0.0F, 1.0F, 0.0F);
+			//var1.func_35835_b(240);
+			var1.addVertexWithUV((double)(var14 - var3 * var13 - var6 * var13), (double)(var15 - var4 * var13), (double)(var16 - var5 * var13 - var7 * var13), (double)var10, (double)var12);
+			var1.addVertexWithUV((double)(var14 - var3 * var13 + var6 * var13), (double)(var15 + var4 * var13), (double)(var16 - var5 * var13 + var7 * var13), (double)var10, (double)var11);
+			var1.addVertexWithUV((double)(var14 + var3 * var13 + var6 * var13), (double)(var15 + var4 * var13), (double)(var16 + var5 * var13 + var7 * var13), (double)var9, (double)var11);
+			var1.addVertexWithUV((double)(var14 + var3 * var13 - var6 * var13), (double)(var15 - var4 * var13), (double)(var16 + var5 * var13 - var7 * var13), (double)var9, (double)var12);
+			var1.draw();
+			GL11.glPolygonOffset(0.0F, 0.0F);
+			GL11.glEnable(GL11.GL_LIGHTING);
+		}
+	}
+
+	public int func_35115_a(float var1) {
+		return '\uf0f0';
+	}
+
+	public void onUpdate() {
+		this.prevPosX = this.posX;
+		this.prevPosY = this.posY;
+		this.prevPosZ = this.posZ;
+		++this.field_35130_a;
+		if(this.field_35130_a == this.field_35129_ay) {
+			this.setEntityDead();
+		}
+
+	}
+
+	public int getFXLayer() {
+		return 3;
+	}
+}
--- /dev/null
+++ net/minecraft/src/EntityHugeExplodeFX.java
@@ -1,0 +1,33 @@
+package net.minecraft.src;
+
+public class EntityHugeExplodeFX extends EntityFX {
+	private int field_35139_a = 0;
+	private int field_35138_ay = 0;
+
+	public EntityHugeExplodeFX(World var1, double var2, double var4, double var6, double var8, double var10, double var12) {
+		super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
+		this.field_35138_ay = 8;
+	}
+
+	public void renderParticle(Tessellator var1, float var2, float var3, float var4, float var5, float var6, float var7) {
+	}
+
+	public void onUpdate() {
+		for(int var1 = 0; var1 < 6; ++var1) {
+			double var2 = this.posX + (this.rand.nextDouble() - this.rand.nextDouble()) * 4.0D;
+			double var4 = this.posY + (this.rand.nextDouble() - this.rand.nextDouble()) * 4.0D;
+			double var6 = this.posZ + (this.rand.nextDouble() - this.rand.nextDouble()) * 4.0D;
+			this.worldObj.spawnParticle("largeexplode", var2, var4, var6, (double)((float)this.field_35139_a / (float)this.field_35138_ay), 0.0D, 0.0D);
+		}
+
+		++this.field_35139_a;
+		if(this.field_35139_a == this.field_35138_ay) {
+			this.setEntityDead();
+		}
+
+	}
+
+	public int getFXLayer() {
+		return 1;
+	}
+}
--- net/minecraft/src/Explosion.java
+++ net/minecraft/src/Explosion.java
@@ -125,6 +125,7 @@
 
 	public void doExplosionB(boolean var1) {
 		this.worldObj.playSoundEffect(this.explosionX, this.explosionY, this.explosionZ, "random.explode", 4.0F, (1.0F + (this.worldObj.rand.nextFloat() - this.worldObj.rand.nextFloat()) * 0.2F) * 0.7F);
+		this.worldObj.spawnParticle("hugeexplosion", this.explosionX, this.explosionY, this.explosionZ, 0.0D, 0.0D, 0.0D);
 		ArrayList var2 = new ArrayList();
 		var2.addAll(this.destroyedBlockPositions);
 
--- net/minecraft/src/RenderGlobal.java
+++ net/minecraft/src/RenderGlobal.java
@@ -1279,6 +1279,14 @@
 			double var16 = this.mc.renderViewEntity.posY - var4;
 			double var18 = this.mc.renderViewEntity.posZ - var6;
 			double var20 = 16.0D;
+			if(var1.equals("hugeexplosion")) {
+				this.mc.effectRenderer.addEffect(new EntityHugeExplodeFX(this.worldObj, var2, var4, var6, var8, var10, var12));
+			}
+
+			if(var1.equals("largeexplode")) {
+				this.mc.effectRenderer.addEffect(new EntityLargeExplodeFX(this.renderEngine, this.worldObj, var2, var4, var6, var8, var10, var12));
+			}
+			
 			if(var14 * var14 + var16 * var16 + var18 * var18 <= var20 * var20) {
 				if(var1.equals("bubble")) {
 					this.mc.effectRenderer.addEffect(new EntityBubbleFX(this.worldObj, var2, var4, var6, var8, var10, var12));
