--- net/minecraft/src/RenderFireball.java
+++ net/minecraft/src/RenderFireball.java
@@ -10,7 +10,7 @@
 		GL11.glEnable(GL12.GL_RESCALE_NORMAL);
 		float var10 = 2.0F;
 		GL11.glScalef(var10 / 1.0F, var10 / 1.0F, var10 / 1.0F);
-		int var11 = Item.snowball.getIconFromDamage(0);
+		int var11 = 46;
 		this.loadTexture("/gui/items.png");
 		Tessellator var12 = Tessellator.instance;
 		float var13 = (float)(var11 % 16 * 16 + 0) / 256.0F;
