--- net/minecraft/src/MusInputStream.java
+++ net/minecraft/src/MusInputStream.java
@@ -15,7 +15,7 @@
 		this.buffer = new byte[1];
 		this.inputStream = var3;
 		String var4 = var2.getPath();
-		var4 = var4.substring(var4.lastIndexOf("/") + 1);
+		var4 = var4.substring(var4.lastIndexOf("/") + 1).replaceAll("%20", " ");
 		this.hash = var4.hashCode();
 	}
 
--- net/minecraft/src/ThreadDownloadResources.java
+++ net/minecraft/src/ThreadDownloadResources.java
@@ -16,6 +16,7 @@
 
 public class ThreadDownloadResources extends Thread {
 	public File resourcesFolder;
+	public File customFolder;
 	private Minecraft mc;
 	private boolean closing = false;
 
@@ -24,12 +25,32 @@
 		this.setName("Resource download thread");
 		this.setDaemon(true);
 		this.resourcesFolder = new File(var1, "resources/");
+		this.customFolder = new File(var1, "custom_resources/");
 		if(!this.resourcesFolder.exists() && !this.resourcesFolder.mkdirs()) {
 			throw new RuntimeException("The working directory could not be created: " + this.resourcesFolder);
 		}
+
+		if(!this.customFolder.exists() && !this.customFolder.mkdirs()) {
+			throw new RuntimeException("The working directory could not be created: " + this.customFolder);
+		}
+
+		try {
+			String[] folderNames = new String[]{"sound", "newsound", "streaming", "music", "newmusic"};
+			File file;
+			for(int i = 0; i < folderNames.length; ++i) {
+				file = new File(this.customFolder, folderNames[i]);
+				if(!file.exists() && !file.mkdirs()) {
+					throw new RuntimeException("The working directory could not be created: " + file);
+				}
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
 	}
 
 	public void run() {
+		this.loadResource(this.customFolder, "");
+
 		try {
 			URL var1 = new URL("http://s3.amazonaws.com/MinecraftResources/");
 			DocumentBuilderFactory var2 = DocumentBuilderFactory.newInstance();
@@ -61,6 +82,7 @@
 	}
 
 	public void reloadResources() {
+		this.loadResource(this.customFolder, "");
 		this.loadResource(this.resourcesFolder, "");
 	}
 
@@ -127,6 +149,8 @@
 			var7.write(var5, 0, var9);
 		} while(!this.closing);
 
+		var6.close();
+		var7.close();
 	}
 
 	public void closeMinecraft() {
--- net/minecraft/src/SoundPool.java
+++ net/minecraft/src/SoundPool.java
@@ -12,11 +12,16 @@
 	private Random rand = new Random();
 	private Map nameToSoundPoolEntriesMapping = new HashMap();
 	private List allSoundPoolEntries = new ArrayList();
+	public List allSoundPoolSoundNames = new ArrayList();
 	public int numberOfSoundPoolEntries = 0;
 	public boolean field_1657_b = true;
 
 	public SoundPoolEntry addSound(String var1, File var2) {
 		try {
+			if(this.allSoundPoolSoundNames.contains(var1)) {
+				return null;
+			}
+
 			String var3 = var1;
 			var1 = var1.substring(0, var1.indexOf("."));
 			if(this.field_1657_b) {
@@ -33,6 +38,7 @@
 			SoundPoolEntry var4 = new SoundPoolEntry(var3, var2.toURI().toURL());
 			((List)this.nameToSoundPoolEntriesMapping.get(var1)).add(var4);
 			this.allSoundPoolEntries.add(var4);
+			this.allSoundPoolSoundNames.add(var3);
 			++this.numberOfSoundPoolEntries;
 			return var4;
 		} catch (MalformedURLException var5) {
