--- net/minecraft/src/EntityFish.java
+++ net/minecraft/src/EntityFish.java
@@ -352,7 +352,7 @@
 			double var8 = (double)MathHelper.sqrt_double(var2 * var2 + var4 * var4 + var6 * var6);
 			double var10 = 0.1D;
 			this.bobber.motionX += var2 * var10;
-			this.bobber.motionY += var4 * var10 + (double)MathHelper.sqrt_double(var8) * 0.08D;
+			this.bobber.motionY += var4 * var10 + (double)MathHelper.sqrt_double(var8) * 0.04D;
 			this.bobber.motionZ += var6 * var10;
 			var1 = 3;
 		} else if(this.ticksCatchable > 0) {
@@ -363,7 +363,7 @@
 			double var9 = (double)MathHelper.sqrt_double(var3 * var3 + var5 * var5 + var7 * var7);
 			double var11 = 0.1D;
 			var13.motionX = var3 * var11;
-			var13.motionY = var5 * var11 + (double)MathHelper.sqrt_double(var9) * 0.08D;
+			var13.motionY = var5 * var11 + (double)MathHelper.sqrt_double(var9) * 0.04D;
 			var13.motionZ = var7 * var11;
 			this.worldObj.entityJoinedWorld(var13);
 			this.angler.addStat(StatList.fishCaughtStat, 1);
