--- net/minecraft/src/BlockLeaves.java
+++ net/minecraft/src/BlockLeaves.java
@@ -156,6 +156,14 @@
 		return Block.sapling.blockID;
 	}
 
+	public void dropBlockAsItemWithChance(World var1, int var2, int var3, int var4, int var5, float var6) {
+		super.dropBlockAsItemWithChance(var1, var2, var3, var4, var5, var6);
+		if(!var1.multiplayerWorld && (var5 & 3) == 0 && var1.rand.nextInt(200) == 0) {
+			this.dropBlockAsItem_do(var1, var2, var3, var4, new ItemStack(Item.appleRed, 1, 0));
+		}
+
+	}
+
 	public void harvestBlock(World var1, EntityPlayer var2, int var3, int var4, int var5, int var6) {
 		if(!var1.multiplayerWorld && var2.getCurrentEquippedItem() != null && var2.getCurrentEquippedItem().itemID == Item.shears.shiftedIndex) {
 			var2.addStat(StatList.mineBlockStatArray[this.blockID], 1);
