--- net/minecraft/client/Minecraft.java
+++ net/minecraft/client/Minecraft.java
@@ -1144,6 +1144,10 @@
 					--this.theWorld.field_27172_i;
 				}
 
+				if(this.thePlayer.isDead) {
+					++this.thePlayer.deathTime;
+				}
+
 				this.theWorld.updateEntities();
 			}
 
--- net/minecraft/src/ItemRenderer.java
+++ net/minecraft/src/ItemRenderer.java
@@ -425,6 +425,10 @@
 
 		float var5 = 0.4F;
 		float var6 = var4 ? 1.0F : 0.0F;
+		if(var1.health <= 0) {
+			var6 = 0.0F;
+		}
+		
 		float var7 = var6 - this.equippedProgress;
 		if(var7 < -var5) {
 			var7 = -var5;
