--- net/minecraft/src/ItemFood.java
+++ net/minecraft/src/ItemFood.java
@@ -12,8 +12,12 @@
 	}
 
 	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
+		if(var3.health >= 20) {
+			return var1;
+		}
+
 		--var1.stackSize;
-		var3.heal(this.healAmount);
+		var3.healAmount += this.healAmount;
 		return var1;
 	}
 
--- net/minecraft/src/EntityLiving.java
+++ net/minecraft/src/EntityLiving.java
@@ -56,6 +56,7 @@
 	protected float moveSpeed = 0.7F;
 	private Entity currentTarget;
 	protected int numTicksToChaseTarget = 0;
+	public int healAmount = 0;
 
 	public EntityLiving(World var1) {
 		super(var1);
@@ -581,6 +582,10 @@
 	}
 
 	public void onLivingUpdate() {
+		if(this.healAmount > 0) {
+			this.heal(this.healAmount);
+			this.healAmount = 0;
+		}
 		if(this.newPosRotationIncrements > 0) {
 			double var1 = this.posX + (this.newPosX - this.posX) / (double)this.newPosRotationIncrements;
 			double var3 = this.posY + (this.newPosY - this.posY) / (double)this.newPosRotationIncrements;
