--- net/minecraft/src/World.java
+++ net/minecraft/src/World.java
@@ -613,6 +613,31 @@
 		}
 	}
 
+	public int getSkyBlockTypeBrightness(EnumSkyBlock var1, int var2, int var3, int var4) {
+		if(this.worldProvider.hasNoSky && var1 == EnumSkyBlock.Sky) {
+			return 0;
+		} else {
+			if(var3 < 0) {
+				var3 = 0;
+			}
+
+			if(var3 >= 128 && var1 == EnumSkyBlock.Sky) {
+				return 15;
+			} else if(var3 >= 0 && var3 < 128 && var2 >= -30000000 && var4 >= -30000000 && var2 < 30000000 && var4 < 30000000) {
+				int var5 = var2 >> 4;
+				int var6 = var4 >> 4;
+				if(!this.chunkExists(var5, var6)) {
+					return 0;
+				} else {
+					Chunk var7 = this.getChunkFromChunkCoords(var5, var6);
+					return var7.getSavedLightValue(var1, var2 & 15, var3, var4 & 15);
+				}
+			} else {
+				return var1.field_1722_c;
+			}
+		}
+	}
+
 	public void neighborLightPropagationChanged(EnumSkyBlock var1, int var2, int var3, int var4, int var5) {
 		if(!this.worldProvider.hasNoSky || var1 != EnumSkyBlock.Sky) {
 			if(this.blockExists(var2, var3, var4)) {
@@ -676,6 +701,16 @@
 		}
 	}
 
+	public int getLightBrightnessForSkyBlocks(int var1, int var2, int var3, int var4) {
+		int var5 = this.getSkyBlockTypeBrightness(EnumSkyBlock.Sky, var1, var2, var3);
+		int var6 = this.getSkyBlockTypeBrightness(EnumSkyBlock.Block, var1, var2, var3);
+		if(var6 < var4) {
+			var6 = var4;
+		}
+
+		return var5 << 20 | var6 << 4;
+	}
+
 	public float getBrightness(int var1, int var2, int var3, int var4) {
 		int var5 = this.getBlockLightValue(var1, var2, var3);
 		if(var5 < var4) {
@@ -1066,6 +1101,15 @@
 		return this.worldProvider.calculateCelestialAngle(this.worldInfo.getWorldTime(), var1);
 	}
 
+	public int func_40475_d(float var1) {
+		return this.worldProvider.func_40470_b(this.worldInfo.getWorldTime(), var1);
+	}
+
+	public float getCelestialAngleRadians(float var1) {
+		float var2 = this.getCelestialAngle(var1);
+		return var2 * (float)Math.PI * 2.0F;
+	}
+
 	public Vec3D func_628_d(float var1) {
 		float var2 = this.getCelestialAngle(var1);
 		float var3 = MathHelper.cos(var2 * (float)Math.PI * 2.0F) * 2.0F + 0.5F;
@@ -1996,6 +2040,10 @@
 			int var8 = var2 + this.rand.nextInt(var4) - this.rand.nextInt(var4);
 			int var9 = var3 + this.rand.nextInt(var4) - this.rand.nextInt(var4);
 			int var10 = this.getBlockId(var7, var8, var9);
+			if(this.rand.nextInt(8) > var8 && var10 == 0) {
+				this.spawnParticle("depthsuspend", (double)((float)var7 + this.rand.nextFloat()), (double)((float)var8 + this.rand.nextFloat()), (double)((float)var9 + this.rand.nextFloat()), 0.0D, 0.0D, 0.0D);
+			}
+			
 			if(var10 > 0) {
 				Block.blocksList[var10].randomDisplayTick(this, var7, var8, var9, var5);
 			}
--- /dev/null
+++ net/minecraft/src/EntityAuraFX.java
@@ -1,0 +1,33 @@
+package net.minecraft.src;
+
+public class EntityAuraFX extends EntityFX {
+	public EntityAuraFX(World var1, double var2, double var4, double var6, double var8, double var10, double var12) {
+		super(var1, var2, var4, var6, var8, var10, var12);
+		float var14 = this.rand.nextFloat() * 0.1F + 0.2F;
+		this.particleRed = var14;
+		this.particleGreen = var14;
+		this.particleBlue = var14;
+		this.particleTextureIndex = 0;
+		this.setSize(0.02F, 0.02F);
+		this.particleScale *= this.rand.nextFloat() * 0.6F + 0.5F;
+		this.motionX *= (double)0.02F;
+		this.motionY *= (double)0.02F;
+		this.motionZ *= (double)0.02F;
+		this.particleMaxAge = (int)(20.0D / (Math.random() * 0.8D + 0.2D));
+		this.noClip = true;
+	}
+
+	public void onUpdate() {
+		this.prevPosX = this.posX;
+		this.prevPosY = this.posY;
+		this.prevPosZ = this.posZ;
+		this.moveEntity(this.motionX, this.motionY, this.motionZ);
+		this.motionX *= 0.99D;
+		this.motionY *= 0.99D;
+		this.motionZ *= 0.99D;
+		if(this.particleMaxAge-- <= 0) {
+			this.setEntityDead();
+		}
+
+	}
+}
--- net/minecraft/src/EntityRenderer.java
+++ net/minecraft/src/EntityRenderer.java
@@ -137,7 +137,8 @@
 	private float getFOVModifier(float var1) {
 		EntityLiving var2 = this.mc.renderViewEntity;
 		float var3 = 70.0F;
-		if(var2.isInsideOfMaterial(Material.water)) {
+		int var6 = ActiveRenderInfo.func_41066_a(this.mc.theWorld, var2, var1);
+		if(var6 != 0 && Block.blocksList[var6].blockMaterial == Material.water) {
 			var3 = 60.0F;
 		}
 
@@ -467,6 +468,7 @@
 			GL11.glClear(GL11.GL_DEPTH_BUFFER_BIT | GL11.GL_COLOR_BUFFER_BIT);
 			GL11.glEnable(GL11.GL_CULL_FACE);
 			this.setupCameraTransform(var1, var18);
+			ActiveRenderInfo.func_41067_a(this.mc.thePlayer, this.mc.gameSettings.thirdPersonView);
 			ClippingHelperImpl.getInstance();
 			if(this.mc.gameSettings.renderDistance < 2) {
 				this.setupFog(-1, var1);
@@ -798,11 +800,29 @@
 		this.fogColorRed = (float)var9.xCoord;
 		this.fogColorGreen = (float)var9.yCoord;
 		this.fogColorBlue = (float)var9.zCoord;
+		float var11;
+		if(this.mc.gameSettings.renderDistance < 2) {
+			Vec3D var10 = MathHelper.sin(var2.getCelestialAngleRadians(var1)) > 0.0F ? Vec3D.createVector(-1.0D, 0.0D, 0.0D) : Vec3D.createVector(1.0D, 0.0D, 0.0D);
+			var11 = (float)var3.getLook(var1).dotProduct(var10);
+			if(var11 < 0.0F) {
+				var11 = 0.0F;
+			}
+
+			if(var11 > 0.0F) {
+				float[] var60 = var2.worldProvider.calcSunriseSunsetColors(var2.getCelestialAngle(var1), var1);
+				if(var60 != null) {
+					var11 *= var60[3];
+					this.fogColorRed = this.fogColorRed * (1.0F - var11) + var60[0] * var11;
+					this.fogColorGreen = this.fogColorGreen * (1.0F - var11) + var60[1] * var11;
+					this.fogColorBlue = this.fogColorBlue * (1.0F - var11) + var60[2] * var11;
+				}
+			}
+		}
+
 		this.fogColorRed += (var6 - this.fogColorRed) * var4;
 		this.fogColorGreen += (var7 - this.fogColorGreen) * var4;
 		this.fogColorBlue += (var8 - this.fogColorBlue) * var4;
 		float var10 = var2.func_27162_g(var1);
-		float var11;
 		float var12;
 		if(var10 > 0.0F) {
 			var11 = 1.0F - var10 * 0.5F;
@@ -820,16 +840,17 @@
 			this.fogColorBlue *= var12;
 		}
 
+		int var21 = ActiveRenderInfo.func_41066_a(this.mc.theWorld, var3, var1);
 		if(this.cloudFog) {
 			Vec3D var16 = var2.func_628_d(var1);
 			this.fogColorRed = (float)var16.xCoord;
 			this.fogColorGreen = (float)var16.yCoord;
 			this.fogColorBlue = (float)var16.zCoord;
-		} else if(var3.isInsideOfMaterial(Material.water)) {
+		} else if(var21 != 0 && Block.blocksList[var21].blockMaterial == Material.water) {
 			this.fogColorRed = 0.02F;
 			this.fogColorGreen = 0.02F;
 			this.fogColorBlue = 0.2F;
-		} else if(var3.isInsideOfMaterial(Material.lava)) {
+		} else if(var21 != 0 && Block.blocksList[var21].blockMaterial == Material.lava) {
 			this.fogColorRed = 0.6F;
 			this.fogColorGreen = 0.1F;
 			this.fogColorBlue = 0.0F;
@@ -839,6 +860,18 @@
 		this.fogColorRed *= var12;
 		this.fogColorGreen *= var12;
 		this.fogColorBlue *= var12;
+		double var58 = (var3.lastTickPosY + (var3.posY - var3.lastTickPosY) * (double)var1) / 32.0D;
+		if(var58 < 1.0D) {
+			if(var58 < 0.0D) {
+				var58 = 0.0D;
+			}
+
+			var58 *= var58;
+			this.fogColorRed = (float)((double)this.fogColorRed * var58);
+			this.fogColorGreen = (float)((double)this.fogColorGreen * var58);
+			this.fogColorBlue = (float)((double)this.fogColorBlue * var58);
+		}
+		
 		if(this.mc.gameSettings.anaglyph) {
 			float var13 = (this.fogColorRed * 30.0F + this.fogColorGreen * 59.0F + this.fogColorBlue * 11.0F) / 100.0F;
 			float var14 = (this.fogColorRed * 30.0F + this.fogColorGreen * 70.0F) / 100.0F;
@@ -853,63 +886,96 @@
 
 	private void setupFog(int var1, float var2) {
 		EntityLiving var3 = this.mc.renderViewEntity;
-		GL11.glFog(GL11.GL_FOG_COLOR, this.func_908_a(this.fogColorRed, this.fogColorGreen, this.fogColorBlue, 1.0F));
-		GL11.glNormal3f(0.0F, -1.0F, 0.0F);
-		GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
-		float var4;
-		float var5;
-		float var6;
-		float var7;
-		float var8;
-		float var9;
-		if(this.cloudFog) {
-			GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
-			GL11.glFogf(GL11.GL_FOG_DENSITY, 0.1F);
-			var4 = 1.0F;
-			var5 = 1.0F;
-			var6 = 1.0F;
-			if(this.mc.gameSettings.anaglyph) {
-				var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
-				var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
-				var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
-			}
-		} else if(var3.isInsideOfMaterial(Material.water)) {
-			GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
-			GL11.glFogf(GL11.GL_FOG_DENSITY, 0.1F);
-			var4 = 0.4F;
-			var5 = 0.4F;
-			var6 = 0.9F;
-			if(this.mc.gameSettings.anaglyph) {
-				var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
-				var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
-				var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
-			}
-		} else if(var3.isInsideOfMaterial(Material.lava)) {
-			GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
-			GL11.glFogf(GL11.GL_FOG_DENSITY, 2.0F);
-			var4 = 0.4F;
-			var5 = 0.3F;
-			var6 = 0.3F;
-			if(this.mc.gameSettings.anaglyph) {
-				var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
-				var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
-				var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
-			}
-		} else {
+		if(var1 == 999) {
+			GL11.glFog(GL11.GL_FOG_COLOR, this.func_908_a(0.0F, 0.0F, 0.0F, 1.0F));
 			GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_LINEAR);
-			GL11.glFogf(GL11.GL_FOG_START, this.farPlaneDistance * 0.25F);
-			GL11.glFogf(GL11.GL_FOG_END, this.farPlaneDistance);
-			if(var1 < 0) {
-				GL11.glFogf(GL11.GL_FOG_START, 0.0F);
-				GL11.glFogf(GL11.GL_FOG_END, this.farPlaneDistance * 0.8F);
-			}
-
+			GL11.glFogf(GL11.GL_FOG_START, 0.0F);
+			GL11.glFogf(GL11.GL_FOG_END, 8.0F);
 			if(GLContext.getCapabilities().GL_NV_fog_distance) {
 				GL11.glFogi(NVFogDistance.GL_FOG_DISTANCE_MODE_NV, NVFogDistance.GL_EYE_RADIAL_NV);
 			}
 
-			if(this.mc.theWorld.worldProvider.isNether) {
-				GL11.glFogf(GL11.GL_FOG_START, 0.0F);
+			GL11.glFogf(GL11.GL_FOG_START, 0.0F);
+		} else {
+			GL11.glFog(GL11.GL_FOG_COLOR, this.func_908_a(this.fogColorRed, this.fogColorGreen, this.fogColorBlue, 1.0F));
+			GL11.glNormal3f(0.0F, -1.0F, 0.0F);
+			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
+			int var70 = ActiveRenderInfo.func_41066_a(this.mc.theWorld, var3, var2);
+			float var4;
+			float var5;
+			float var6;
+			float var7;
+			float var8;
+			float var9;
+			if(this.cloudFog) {
+				GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
+				GL11.glFogf(GL11.GL_FOG_DENSITY, 0.1F);
+				var4 = 1.0F;
+				var5 = 1.0F;
+				var6 = 1.0F;
+				if(this.mc.gameSettings.anaglyph) {
+					var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
+					var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
+					var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
+				}
+			} else if(var70 > 0 && Block.blocksList[var70].blockMaterial == Material.water) {
+				GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
+				GL11.glFogf(GL11.GL_FOG_DENSITY, 0.1F);
+				var4 = 0.4F;
+				var5 = 0.4F;
+				var6 = 0.9F;
+				if(this.mc.gameSettings.anaglyph) {
+					var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
+					var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
+					var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
+				}
+			} else if(var70 > 0 && Block.blocksList[var70].blockMaterial == Material.lava) {
+				GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_EXP);
+				GL11.glFogf(GL11.GL_FOG_DENSITY, 2.0F);
+				var4 = 0.4F;
+				var5 = 0.3F;
+				var6 = 0.3F;
+				if(this.mc.gameSettings.anaglyph) {
+					var7 = (var4 * 30.0F + var5 * 59.0F + var6 * 11.0F) / 100.0F;
+					var8 = (var4 * 30.0F + var5 * 70.0F) / 100.0F;
+					var9 = (var4 * 30.0F + var6 * 70.0F) / 100.0F;
+				}
+			} else {
+				if(!this.mc.theWorld.worldProvider.hasNoSky) {
+					double var12 = (double)((var3.getEntityBrightnessForRender(var2) & 15728640) >> 20) / 16.0D + (var3.lastTickPosY + (var3.posY - var3.lastTickPosY) * (double)var2 + 4.0D) / 32.0D;
+					if(var12 < 1.0D) {
+						if(var12 < 0.0D) {
+							var12 = 0.0D;
+						}
+
+						var12 *= var12;
+						var8 = 100.0F * (float)var12;
+						if(var8 < 5.0F) {
+							var8 = 5.0F;
+						}
+
+						if(this.farPlaneDistance > var8) {
+							this.farPlaneDistance = var8;
+						}
+					}
+				}
+
+				GL11.glFogi(GL11.GL_FOG_MODE, GL11.GL_LINEAR);
+				if(var1 < 0) {
+					GL11.glFogf(GL11.GL_FOG_START, 0.0F);
+					GL11.glFogf(GL11.GL_FOG_END, this.farPlaneDistance * 0.8F);
+				} else {
+					GL11.glFogf(GL11.GL_FOG_START, this.farPlaneDistance * 0.25F);
+					GL11.glFogf(GL11.GL_FOG_END, this.farPlaneDistance);
+				}
+
+				if(GLContext.getCapabilities().GL_NV_fog_distance) {
+					GL11.glFogi(NVFogDistance.GL_FOG_DISTANCE_MODE_NV, NVFogDistance.GL_EYE_RADIAL_NV);
+				}
+
+				if(this.mc.theWorld.worldProvider.isNether) {
+					GL11.glFogf(GL11.GL_FOG_START, 0.0F);
+				}
 			}
 		}
 
--- net/minecraft/src/WorldProvider.java
+++ net/minecraft/src/WorldProvider.java
@@ -56,6 +56,10 @@
 		return var5;
 	}
 
+	public int func_40470_b(long var1, float var3) {
+		return (int)(var1 / 24000L) % 8;
+	}
+
 	public float[] calcSunriseSunsetColors(float var1, float var2) {
 		float var3 = 0.4F;
 		float var4 = MathHelper.cos(var1 * (float)Math.PI * 2.0F) - 0.0F;
--- net/minecraft/src/RenderGlobal.java
+++ net/minecraft/src/RenderGlobal.java
@@ -632,6 +632,8 @@
 			float var10;
 			float var11;
 			float var12;
+			float var16;
+			int var20;
 			if(var18 != null) {
 				GL11.glDisable(GL11.GL_TEXTURE_2D);
 				GL11.glShadeModel(GL11.GL_SMOOTH);
@@ -658,10 +660,10 @@
 				byte var19 = 16;
 				var17.setColorRGBA_F(var18[0], var18[1], var18[2], 0.0F);
 
-				for(int var20 = 0; var20 <= var19; ++var20) {
+				for(var20 = 0; var20 <= var19; ++var20) {
 					var14 = (float)var20 * (float)Math.PI * 2.0F / (float)var19;
 					float var15 = MathHelper.sin(var14);
-					float var16 = MathHelper.cos(var14);
+					var16 = MathHelper.cos(var14);
 					var17.addVertex((double)(var15 * 120.0F), (double)(var16 * 120.0F), (double)(-var16 * 40.0F * var18[3]));
 				}
 
@@ -679,7 +681,7 @@
 			var10 = 0.0F;
 			GL11.glColor4f(1.0F, 1.0F, 1.0F, var7);
 			GL11.glTranslatef(var8, var9, var10);
-			GL11.glRotatef(0.0F, 0.0F, 0.0F, 1.0F);
+			GL11.glRotatef(-90.0F, 0.0F, 1.0F, 0.0F);
 			GL11.glRotatef(this.worldObj.getCelestialAngle(var1) * 360.0F, 1.0F, 0.0F, 0.0F);
 			var11 = 30.0F;
 			GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.renderEngine.getTexture("/terrain/sun.png"));
@@ -690,12 +692,19 @@
 			var17.addVertexWithUV((double)(-var11), 100.0D, (double)var11, 0.0D, 1.0D);
 			var17.draw();
 			var11 = 20.0F;
-			GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.renderEngine.getTexture("/terrain/moon.png"));
+			GL11.glBindTexture(GL11.GL_TEXTURE_2D, this.renderEngine.getTexture("/terrain/moon_phases.png"));
+			var20 = this.worldObj.func_40475_d(var1);
+			int var26 = var20 % 4;
+			int var27 = var20 / 4 % 2;
+			var16 = (float)(var26 + 0) / 4.0F;
+			float var21 = (float)(var27 + 0) / 2.0F;
+			float var22 = (float)(var26 + 1) / 4.0F;
+			float var23 = (float)(var27 + 1) / 2.0F;
 			var17.startDrawingQuads();
-			var17.addVertexWithUV((double)(-var11), -100.0D, (double)var11, 1.0D, 1.0D);
-			var17.addVertexWithUV((double)var11, -100.0D, (double)var11, 0.0D, 1.0D);
-			var17.addVertexWithUV((double)var11, -100.0D, (double)(-var11), 0.0D, 0.0D);
-			var17.addVertexWithUV((double)(-var11), -100.0D, (double)(-var11), 1.0D, 0.0D);
+			var17.addVertexWithUV((double)(-var11), -100.0D, (double)var11, (double)var22, (double)var23);
+			var17.addVertexWithUV((double)var11, -100.0D, (double)var11, (double)var16, (double)var23);
+			var17.addVertexWithUV((double)var11, -100.0D, (double)(-var11), (double)var16, (double)var21);
+			var17.addVertexWithUV((double)(-var11), -100.0D, (double)(-var11), (double)var22, (double)var21);
 			var17.draw();
 			GL11.glDisable(GL11.GL_TEXTURE_2D);
 			var12 = this.worldObj.getStarBrightness(var1) * var7;
@@ -709,14 +718,52 @@
 			GL11.glEnable(GL11.GL_ALPHA_TEST);
 			GL11.glEnable(GL11.GL_FOG);
 			GL11.glPopMatrix();
+			GL11.glDisable(GL11.GL_TEXTURE_2D);
+			GL11.glColor3f(0.0F, 0.0F, 0.0F);
+			double var30 = this.mc.thePlayer.getPosition(var1).yCoord - 63.0D;
+			if(var30 < 0.0D) {
+				GL11.glPushMatrix();
+				GL11.glTranslatef(0.0F, 12.0F, 0.0F);
+				GL11.glCallList(this.glSkyList2);
+				GL11.glPopMatrix();
+				var9 = 1.0F;
+				var10 = -((float)(var30 + 65.0D));
+				var11 = -var9;
+				var17.startDrawingQuads();
+				var17.setColorRGBA_I(0, 255);
+				var17.addVertex((double)(-var9), (double)var10, (double)var9);
+				var17.addVertex((double)var9, (double)var10, (double)var9);
+				var17.addVertex((double)var9, (double)var11, (double)var9);
+				var17.addVertex((double)(-var9), (double)var11, (double)var9);
+				var17.addVertex((double)(-var9), (double)var11, (double)(-var9));
+				var17.addVertex((double)var9, (double)var11, (double)(-var9));
+				var17.addVertex((double)var9, (double)var10, (double)(-var9));
+				var17.addVertex((double)(-var9), (double)var10, (double)(-var9));
+				var17.addVertex((double)var9, (double)var11, (double)(-var9));
+				var17.addVertex((double)var9, (double)var11, (double)var9);
+				var17.addVertex((double)var9, (double)var10, (double)var9);
+				var17.addVertex((double)var9, (double)var10, (double)(-var9));
+				var17.addVertex((double)(-var9), (double)var10, (double)(-var9));
+				var17.addVertex((double)(-var9), (double)var10, (double)var9);
+				var17.addVertex((double)(-var9), (double)var11, (double)var9);
+				var17.addVertex((double)(-var9), (double)var11, (double)(-var9));
+				var17.addVertex((double)(-var9), (double)var11, (double)(-var9));
+				var17.addVertex((double)(-var9), (double)var11, (double)var9);
+				var17.addVertex((double)var9, (double)var11, (double)var9);
+				var17.addVertex((double)var9, (double)var11, (double)(-var9));
+				var17.draw();
+			}
+
 			if(this.worldObj.worldProvider.func_28112_c()) {
 				GL11.glColor3f(var3 * 0.2F + 0.04F, var4 * 0.2F + 0.04F, var5 * 0.6F + 0.1F);
 			} else {
 				GL11.glColor3f(var3, var4, var5);
 			}
 
-			GL11.glDisable(GL11.GL_TEXTURE_2D);
+			GL11.glPushMatrix();
+			GL11.glTranslatef(0.0F, -((float)(var30 - 16.0D)), 0.0F);
 			GL11.glCallList(this.glSkyList2);
+			GL11.glPopMatrix();
 			GL11.glEnable(GL11.GL_TEXTURE_2D);
 			GL11.glDepthMask(true);
 		}
@@ -750,8 +797,8 @@
 				}
 
 				var10 = 0.5F / 1024.0F;
-				double var22 = this.mc.renderViewEntity.prevPosX + (this.mc.renderViewEntity.posX - this.mc.renderViewEntity.prevPosX) * (double)var1 + (double)(((float)this.cloudOffsetX + var1) * 0.03F);
-				double var13 = this.mc.renderViewEntity.prevPosZ + (this.mc.renderViewEntity.posZ - this.mc.renderViewEntity.prevPosZ) * (double)var1;
+				double var22 = this.mc.renderViewEntity.prevPosX + (this.mc.renderViewEntity.posX - this.mc.renderViewEntity.prevPosX) * (double)var1;
+				double var13 = this.mc.renderViewEntity.prevPosZ + (this.mc.renderViewEntity.posZ - this.mc.renderViewEntity.prevPosZ) * (double)var1 + (double)(((float)this.cloudOffsetX + var1) * 0.03F);
 				int var15 = MathHelper.floor_double(var22 / 2048.0D);
 				int var16 = MathHelper.floor_double(var13 / 2048.0D);
 				var22 -= (double)(var15 * 2048);
@@ -789,8 +836,8 @@
 		Tessellator var3 = Tessellator.instance;
 		float var4 = 12.0F;
 		float var5 = 4.0F;
-		double var6 = (this.mc.renderViewEntity.prevPosX + (this.mc.renderViewEntity.posX - this.mc.renderViewEntity.prevPosX) * (double)var1 + (double)(((float)this.cloudOffsetX + var1) * 0.03F)) / (double)var4;
-		double var8 = (this.mc.renderViewEntity.prevPosZ + (this.mc.renderViewEntity.posZ - this.mc.renderViewEntity.prevPosZ) * (double)var1) / (double)var4 + (double)0.33F;
+		double var6 = (this.mc.renderViewEntity.prevPosX + (this.mc.renderViewEntity.posX - this.mc.renderViewEntity.prevPosX) * (double)var1) / (double)var4 + (double)0.33F;
+		double var8 = (this.mc.renderViewEntity.prevPosZ + (this.mc.renderViewEntity.posZ - this.mc.renderViewEntity.prevPosZ) * (double)var1 + (double)(((float)this.cloudOffsetX + var1) * 0.03F)) / (double)var4;
 		float var10 = this.worldObj.worldProvider.getCloudHeight() - var2 + 0.33F;
 		int var11 = MathHelper.floor_double(var6 / 2048.0D);
 		int var12 = MathHelper.floor_double(var8 / 2048.0D);
@@ -1282,6 +1329,8 @@
 			if(var14 * var14 + var16 * var16 + var18 * var18 <= var20 * var20) {
 				if(var1.equals("bubble")) {
 					this.mc.effectRenderer.addEffect(new EntityBubbleFX(this.worldObj, var2, var4, var6, var8, var10, var12));
+				} else if(var1.equals("depthsuspend")) {
+					this.mc.effectRenderer.addEffect(new EntityAuraFX(this.worldObj, var2, var4, var6, var8, var10, var12));
 				} else if(var1.equals("smoke")) {
 					this.mc.effectRenderer.addEffect(new EntitySmokeFX(this.worldObj, var2, var4, var6, var8, var10, var12));
 				} else if(var1.equals("note")) {
--- net/minecraft/src/EffectRenderer.java
+++ net/minecraft/src/EffectRenderer.java
@@ -47,11 +47,11 @@
 	}
 
 	public void renderParticles(Entity var1, float var2) {
-		float var3 = MathHelper.cos(var1.rotationYaw * (float)Math.PI / 180.0F);
-		float var4 = MathHelper.sin(var1.rotationYaw * (float)Math.PI / 180.0F);
-		float var5 = -var4 * MathHelper.sin(var1.rotationPitch * (float)Math.PI / 180.0F);
-		float var6 = var3 * MathHelper.sin(var1.rotationPitch * (float)Math.PI / 180.0F);
-		float var7 = MathHelper.cos(var1.rotationPitch * (float)Math.PI / 180.0F);
+		float var3 = ActiveRenderInfo.field_41070_d;
+		float var4 = ActiveRenderInfo.field_41068_f;
+		float var5 = ActiveRenderInfo.field_41069_g;
+		float var6 = ActiveRenderInfo.field_41078_h;
+		float var7 = ActiveRenderInfo.field_41071_e;
 		EntityFX.interpPosX = var1.lastTickPosX + (var1.posX - var1.lastTickPosX) * (double)var2;
 		EntityFX.interpPosY = var1.lastTickPosY + (var1.posY - var1.lastTickPosY) * (double)var2;
 		EntityFX.interpPosZ = var1.lastTickPosZ + (var1.posZ - var1.lastTickPosZ) * (double)var2;
--- net/minecraft/src/Entity.java
+++ net/minecraft/src/Entity.java
@@ -584,6 +584,18 @@
 		}
 	}
 
+	public int getEntityBrightnessForRender(float var1) {
+		int var2 = MathHelper.floor_double(this.posX);
+		int var3 = MathHelper.floor_double(this.posZ);
+		if(this.worldObj.blockExists(var2, 128 / 2, var3)) {
+			double var4 = (this.boundingBox.maxY - this.boundingBox.minY) * 0.66D;
+			int var6 = MathHelper.floor_double(this.posY - (double)this.yOffset + var4);
+			return this.worldObj.getLightBrightnessForSkyBlocks(var2, var6, var3, 0);
+		} else {
+			return 0;
+		}
+	}
+
 	public float getEntityBrightness(float var1) {
 		int var2 = MathHelper.floor_double(this.posX);
 		double var3 = (this.boundingBox.maxY - this.boundingBox.minY) * 0.66D;
--- net/minecraft/src/ChunkPosition.java
+++ net/minecraft/src/ChunkPosition.java
@@ -11,6 +11,10 @@
 		this.z = var3;
 	}
 
+	public ChunkPosition(Vec3D var1) {
+		this(MathHelper.floor_double(var1.xCoord), MathHelper.floor_double(var1.yCoord), MathHelper.floor_double(var1.zCoord));
+	}
+
 	public boolean equals(Object var1) {
 		if(!(var1 instanceof ChunkPosition)) {
 			return false;
--- net/minecraft/src/Vec3D.java
+++ net/minecraft/src/Vec3D.java
@@ -65,6 +65,10 @@
 		return var1 < 1.0E-4D ? createVector(0.0D, 0.0D, 0.0D) : createVector(this.xCoord / var1, this.yCoord / var1, this.zCoord / var1);
 	}
 
+	public double dotProduct(Vec3D var1) {
+		return this.xCoord * var1.xCoord + this.yCoord * var1.yCoord + this.zCoord * var1.zCoord;
+	}
+
 	public Vec3D crossProduct(Vec3D var1) {
 		return createVector(this.yCoord * var1.zCoord - this.zCoord * var1.yCoord, this.zCoord * var1.xCoord - this.xCoord * var1.zCoord, this.xCoord * var1.yCoord - this.yCoord * var1.xCoord);
 	}
--- /dev/null
+++ net/minecraft/src/ActiveRenderInfo.java
@@ -1,0 +1,66 @@
+package net.minecraft.src;
+
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+import org.lwjgl.opengl.GL11;
+import org.lwjgl.util.glu.GLU;
+
+public class ActiveRenderInfo {
+	public static float field_41074_a = 0.0F;
+	public static float field_41072_b = 0.0F;
+	public static float field_41073_c = 0.0F;
+	private static IntBuffer field_41079_i = GLAllocation.createDirectIntBuffer(16);
+	private static FloatBuffer field_41076_j = GLAllocation.createDirectFloatBuffer(16);
+	private static FloatBuffer field_41077_k = GLAllocation.createDirectFloatBuffer(16);
+	private static FloatBuffer field_41075_l = GLAllocation.createDirectFloatBuffer(3);
+	public static float field_41070_d;
+	public static float field_41071_e;
+	public static float field_41068_f;
+	public static float field_41069_g;
+	public static float field_41078_h;
+
+	public static void func_41067_a(EntityPlayer var0, boolean var1) {
+		GL11.glGetFloat(GL11.GL_MODELVIEW_MATRIX, field_41076_j);
+		GL11.glGetFloat(GL11.GL_PROJECTION_MATRIX, field_41077_k);
+		GL11.glGetInteger(GL11.GL_VIEWPORT, field_41079_i);
+		float var2 = (float)((field_41079_i.get(0) + field_41079_i.get(2)) / 2);
+		float var3 = (float)((field_41079_i.get(1) + field_41079_i.get(3)) / 2);
+		GLU.gluUnProject(var2, var3, 0.0F, field_41076_j, field_41077_k, field_41079_i, field_41075_l);
+		field_41074_a = field_41075_l.get(0);
+		field_41072_b = field_41075_l.get(1);
+		field_41073_c = field_41075_l.get(2);
+		int var4 = var1 ? 1 : 0;
+		float var5 = var0.rotationPitch;
+		float var6 = var0.rotationYaw;
+		field_41070_d = MathHelper.cos(var6 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
+		field_41068_f = MathHelper.sin(var6 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
+		field_41069_g = -field_41068_f * MathHelper.sin(var5 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
+		field_41078_h = field_41070_d * MathHelper.sin(var5 * (float)Math.PI / 180.0F) * (float)(1 - var4 * 2);
+		field_41071_e = MathHelper.cos(var5 * (float)Math.PI / 180.0F);
+	}
+
+	public static Vec3D func_41065_a(EntityLiving var0, double var1) {
+		double var3 = var0.prevPosX + (var0.posX - var0.prevPosX) * var1;
+		double var5 = var0.prevPosY + (var0.posY - var0.prevPosY) * var1 + (double)var0.getEyeHeight();
+		double var7 = var0.prevPosZ + (var0.posZ - var0.prevPosZ) * var1;
+		double var9 = var3 + (double)(field_41074_a * 1.0F);
+		double var11 = var5 + (double)(field_41072_b * 1.0F);
+		double var13 = var7 + (double)(field_41073_c * 1.0F);
+		return Vec3D.createVector(var9, var11, var13);
+	}
+
+	public static int func_41066_a(World var0, EntityLiving var1, float var2) {
+		Vec3D var3 = func_41065_a(var1, (double)var2);
+		ChunkPosition var4 = new ChunkPosition(var3);
+		int var5 = var0.getBlockId(var4.x, var4.y, var4.z);
+		if(var5 != 0 && Block.blocksList[var5].blockMaterial.getIsLiquid()) {
+			float var6 = BlockFluid.getPercentAir(var0.getBlockMetadata(var4.x, var4.y, var4.z)) - 1.0F / 9.0F;
+			float var7 = (float)(var4.y + 1) - var6;
+			if(var3.yCoord >= (double)var7) {
+				var5 = var0.getBlockId(var4.x, var4.y + 1, var4.z);
+			}
+		}
+
+		return var5;
+	}
+}
