--- net/minecraft/src/EntityDiggingFX.java
+++ net/minecraft/src/EntityDiggingFX.java
@@ -7,7 +7,7 @@
 	public EntityDiggingFX(World var1, double var2, double var4, double var6, double var8, double var10, double var12, Block var14, int var15, int var16) {
 		super(var1, var2, var4, var6, var8, var10, var12);
 		this.field_4082_a = var14;
-		this.particleTextureIndex = var14.getBlockTextureFromSideAndMetadata(0, var16);
+		this.particleTextureIndex = var14.getBlockTextureFromSideAndMetadata(var15, var16);
 		this.particleGravity = var14.blockParticleGravity;
 		this.particleRed = this.particleGreen = this.particleBlue = 0.6F;
 		this.particleScale /= 2.0F;
