--- net/minecraft/src/ModelRenderer.java
+++ net/minecraft/src/ModelRenderer.java
@@ -13,15 +13,22 @@
 	public float rotateAngleX;
 	public float rotateAngleY;
 	public float rotateAngleZ;
+	public float field_35977_i;
+	public float field_35975_j;
+	public float field_35976_k;
+	public float field_35973_l;
+	public float field_35974_m;
+	public float field_35972_n;
 	private boolean compiled = false;
 	private int displayList = 0;
 	public boolean mirror = false;
 	public boolean showModel = true;
 	public boolean field_1402_i = false;
 
-	public ModelRenderer(int var1, int var2) {
-		this.textureOffsetX = var1;
-		this.textureOffsetY = var2;
+	public ModelRenderer(ModelBase var1, int var2, int var3) {
+		this.textureOffsetX = var2;
+		this.textureOffsetY = var3;
+		var1.field_35394_j.add(this);
 	}
 
 	public void addBox(float var1, float var2, float var3, int var4, int var5, int var6) {
@@ -29,6 +36,12 @@
 	}
 
 	public void addBox(float var1, float var2, float var3, int var4, int var5, int var6, float var7) {
+		this.field_35977_i = var1;
+		this.field_35975_j = var2;
+		this.field_35976_k = var3;
+		this.field_35973_l = var1 + (float)var4;
+		this.field_35974_m = var2 + (float)var5;
+		this.field_35972_n = var3 + (float)var6;
 		this.corners = new PositionTextureVertex[8];
 		this.faces = new TexturedQuad[6];
 		float var8 = var1 + (float)var4;
--- net/minecraft/src/ModelChicken.java
+++ net/minecraft/src/ModelChicken.java
@@ -12,28 +12,28 @@
 
 	public ModelChicken() {
 		byte var1 = 16;
-		this.head = new ModelRenderer(0, 0);
+		this.head = new ModelRenderer(this, 0, 0);
 		this.head.addBox(-2.0F, -6.0F, -2.0F, 4, 6, 3, 0.0F);
 		this.head.setRotationPoint(0.0F, (float)(-1 + var1), -4.0F);
-		this.bill = new ModelRenderer(14, 0);
+		this.bill = new ModelRenderer(this, 14, 0);
 		this.bill.addBox(-2.0F, -4.0F, -4.0F, 4, 2, 2, 0.0F);
 		this.bill.setRotationPoint(0.0F, (float)(-1 + var1), -4.0F);
-		this.chin = new ModelRenderer(14, 4);
+		this.chin = new ModelRenderer(this, 14, 4);
 		this.chin.addBox(-1.0F, -2.0F, -3.0F, 2, 2, 2, 0.0F);
 		this.chin.setRotationPoint(0.0F, (float)(-1 + var1), -4.0F);
-		this.body = new ModelRenderer(0, 9);
+		this.body = new ModelRenderer(this, 0, 9);
 		this.body.addBox(-3.0F, -4.0F, -3.0F, 6, 8, 6, 0.0F);
 		this.body.setRotationPoint(0.0F, (float)(0 + var1), 0.0F);
-		this.rightLeg = new ModelRenderer(26, 0);
+		this.rightLeg = new ModelRenderer(this, 26, 0);
 		this.rightLeg.addBox(-1.0F, 0.0F, -3.0F, 3, 5, 3);
 		this.rightLeg.setRotationPoint(-2.0F, (float)(3 + var1), 1.0F);
-		this.leftLeg = new ModelRenderer(26, 0);
+		this.leftLeg = new ModelRenderer(this, 26, 0);
 		this.leftLeg.addBox(-1.0F, 0.0F, -3.0F, 3, 5, 3);
 		this.leftLeg.setRotationPoint(1.0F, (float)(3 + var1), 1.0F);
-		this.rightWing = new ModelRenderer(24, 13);
+		this.rightWing = new ModelRenderer(this, 24, 13);
 		this.rightWing.addBox(0.0F, 0.0F, -3.0F, 1, 4, 6);
 		this.rightWing.setRotationPoint(-4.0F, (float)(-3 + var1), 0.0F);
-		this.leftWing = new ModelRenderer(24, 13);
+		this.leftWing = new ModelRenderer(this, 24, 13);
 		this.leftWing.addBox(-1.0F, 0.0F, -3.0F, 1, 4, 6);
 		this.leftWing.setRotationPoint(4.0F, (float)(-3 + var1), 0.0F);
 	}
--- net/minecraft/src/ModelBoat.java
+++ net/minecraft/src/ModelBoat.java
@@ -4,11 +4,11 @@
 	public ModelRenderer[] boatSides = new ModelRenderer[5];
 
 	public ModelBoat() {
-		this.boatSides[0] = new ModelRenderer(0, 8);
-		this.boatSides[1] = new ModelRenderer(0, 0);
-		this.boatSides[2] = new ModelRenderer(0, 0);
-		this.boatSides[3] = new ModelRenderer(0, 0);
-		this.boatSides[4] = new ModelRenderer(0, 0);
+		this.boatSides[0] = new ModelRenderer(this, 0, 8);
+		this.boatSides[1] = new ModelRenderer(this, 0, 0);
+		this.boatSides[2] = new ModelRenderer(this, 0, 0);
+		this.boatSides[3] = new ModelRenderer(this, 0, 0);
+		this.boatSides[4] = new ModelRenderer(this, 0, 0);
 		byte var1 = 24;
 		byte var2 = 6;
 		byte var3 = 20;
--- net/minecraft/src/ItemFood.java
+++ net/minecraft/src/ItemFood.java
@@ -3,6 +3,7 @@
 public class ItemFood extends Item {
 	private int healAmount;
 	private boolean isWolfsFavoriteMeat;
+	private boolean field_35431_bw;
 
 	public ItemFood(int var1, int var2, boolean var3) {
 		super(var1);
@@ -11,9 +12,27 @@
 		this.maxStackSize = 1;
 	}
 
-	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
+	public ItemStack onFoodEaten(ItemStack var1, World var2, EntityPlayer var3) {
 		--var1.stackSize;
 		var3.heal(this.healAmount);
+		var2.playSoundAtEntity(var3, "random.burp", 0.5F, var2.rand.nextFloat() * 0.1F + 0.9F);
+
+		return var1;
+	}
+
+	public int getMaxItemUseDuration(ItemStack var1) {
+		return 32;
+	}
+
+	public EnumAction getItemUseAction(ItemStack var1) {
+		return EnumAction.eat;
+	}
+
+	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
+		if(var3.health < 20) {
+			var3.setItemInUse(var1, this.getMaxItemUseDuration(var1));
+		}
+
 		return var1;
 	}
 
--- net/minecraft/src/EntityRenderer.java
+++ net/minecraft/src/EntityRenderer.java
@@ -35,6 +35,9 @@
 	private float field_22221_y = 0.0F;
 	private float field_22220_z = 0.0F;
 	private float field_22230_A = 0.0F;
+	private float fovModifierHand;
+	private float fovModifierHandPrev;
+	private float fovMultiplierTemp;
 	private boolean cloudFog = false;
 	private double cameraZoom = 1.0D;
 	private double cameraYaw = 0.0D;
@@ -58,6 +61,7 @@
 	}
 
 	public void updateRenderer() {
+		this.updateFovModifierHand();
 		this.fogColor2 = this.fogColor1;
 		this.field_22227_s = this.field_22228_r;
 		this.field_22225_u = this.field_22226_t;
@@ -134,19 +138,31 @@
 		}
 	}
 
-	private float getFOVModifier(float var1) {
-		EntityLiving var2 = this.mc.renderViewEntity;
-		float var3 = 70.0F;
-		if(var2.isInsideOfMaterial(Material.water)) {
-			var3 = 60.0F;
-		}
-
-		if(var2.health <= 0) {
-			float var4 = (float)var2.deathTime + var1;
-			var3 /= (1.0F - 500.0F / (var4 + 500.0F)) * 2.0F + 1.0F;
-		}
-
-		return var3 + this.field_22221_y + (this.field_22222_x - this.field_22221_y) * var1;
+	private void updateFovModifierHand() {
+		EntityPlayerSP var1 = (EntityPlayerSP)this.mc.renderViewEntity;
+		this.fovMultiplierTemp = var1.getFOVMultiplier();
+		this.fovModifierHandPrev = this.fovModifierHand;
+		this.fovModifierHand += (this.fovMultiplierTemp - this.fovModifierHand) * 0.5F;
+	}
+
+	private float getFOVModifier(float var1, boolean var2) {
+		EntityPlayer var3 = (EntityPlayer)this.mc.renderViewEntity;
+		float var4 = 70.0F;
+		if(var2) {
+			var4 += this.mc.gameSettings.fovSetting * 40.0F;
+			var4 *= this.fovModifierHandPrev + (this.fovModifierHand - this.fovModifierHandPrev) * var1;
+		}
+
+		if(var3.health <= 0) {
+			float var5 = (float)var3.deathTime + var1;
+			var4 /= (1.0F - 500.0F / (var5 + 500.0F)) * 2.0F + 1.0F;
+		}
+
+		if(var3.isInsideOfMaterial(Material.water)) {
+			var4 = var4 * 60.0F / 70.0F;
+		}
+
+		return var4 + this.field_22221_y + (this.field_22222_x - this.field_22221_y) * var1;
 	}
 
 	private void hurtCameraEffect(float var1) {
@@ -270,9 +286,9 @@
 		if(this.cameraZoom != 1.0D) {
 			GL11.glTranslatef((float)this.cameraYaw, (float)(-this.cameraPitch), 0.0F);
 			GL11.glScaled(this.cameraZoom, this.cameraZoom, 1.0D);
-			GLU.gluPerspective(this.getFOVModifier(var1), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
+			GLU.gluPerspective(this.getFOVModifier(var1, true), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
 		} else {
-			GLU.gluPerspective(this.getFOVModifier(var1), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
+			GLU.gluPerspective(this.getFOVModifier(var1, true), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
 		}
 
 		GL11.glMatrixMode(GL11.GL_MODELVIEW);
@@ -299,6 +315,22 @@
 	}
 
 	private void func_4135_b(float var1, int var2) {
+		GL11.glMatrixMode(GL11.GL_PROJECTION);
+		GL11.glLoadIdentity();
+		float var3 = 0.07F;
+		if(this.mc.gameSettings.anaglyph) {
+			GL11.glTranslatef((float)(-(var2 * 2 - 1)) * var3, 0.0F, 0.0F);
+		}
+
+		if(this.cameraZoom != 1.0D) {
+			GL11.glTranslatef((float)this.cameraYaw, (float)(-this.cameraPitch), 0.0F);
+			GL11.glScaled(this.cameraZoom, this.cameraZoom, 1.0D);
+			GLU.gluPerspective(this.getFOVModifier(var1, false), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
+		} else {
+			GLU.gluPerspective(this.getFOVModifier(var1, false), (float)this.mc.displayWidth / (float)this.mc.displayHeight, 0.05F, this.farPlaneDistance * 2.0F);
+		}
+
+		GL11.glMatrixMode(GL11.GL_MODELVIEW);
 		GL11.glLoadIdentity();
 		if(this.mc.gameSettings.anaglyph) {
 			GL11.glTranslatef((float)(var2 * 2 - 1) * 0.1F, 0.0F, 0.0F);
--- net/minecraft/src/EntityArrow.java
+++ net/minecraft/src/EntityArrow.java
@@ -14,6 +14,7 @@
 	public EntityLiving owner;
 	private int ticksInGround;
 	private int ticksInAir = 0;
+	public boolean field_35140_d = false;
 
 	public EntityArrow(World var1) {
 		super(var1);
@@ -27,7 +28,7 @@
 		this.yOffset = 0.0F;
 	}
 
-	public EntityArrow(World var1, EntityLiving var2) {
+	public EntityArrow(World var1, EntityLiving var2, float var3) {
 		super(var1);
 		this.owner = var2;
 		this.doesArrowBelongToPlayer = var2 instanceof EntityPlayer;
@@ -41,7 +42,7 @@
 		this.motionX = (double)(-MathHelper.sin(this.rotationYaw / 180.0F * (float)Math.PI) * MathHelper.cos(this.rotationPitch / 180.0F * (float)Math.PI));
 		this.motionZ = (double)(MathHelper.cos(this.rotationYaw / 180.0F * (float)Math.PI) * MathHelper.cos(this.rotationPitch / 180.0F * (float)Math.PI));
 		this.motionY = (double)(-MathHelper.sin(this.rotationPitch / 180.0F * (float)Math.PI));
-		this.setArrowHeading(this.motionX, this.motionY, this.motionZ, 1.5F, 1.0F);
+		this.setArrowHeading(this.motionX, this.motionY, this.motionZ, var3 * 1.5F, 1.0F);
 	}
 
 	protected void entityInit() {
@@ -136,8 +137,9 @@
 			List var5 = this.worldObj.getEntitiesWithinAABBExcludingEntity(this, this.boundingBox.addCoord(this.motionX, this.motionY, this.motionZ).expand(1.0D, 1.0D, 1.0D));
 			double var6 = 0.0D;
 
+			int var8;
 			float var10;
-			for(int var8 = 0; var8 < var5.size(); ++var8) {
+			for(var8 = 0; var8 < var5.size(); ++var8) {
 				Entity var9 = (Entity)var5.get(var8);
 				if(var9.canBeCollidedWith() && (var9 != this.owner || this.ticksInAir >= 5)) {
 					var10 = 0.3F;
@@ -160,7 +162,17 @@
 			float var19;
 			if(var3 != null) {
 				if(var3.entityHit != null) {
-					if(var3.entityHit.attackEntityFrom(this.owner, 4)) {
+					var19 = MathHelper.sqrt_double(this.motionX * this.motionX + this.motionY * this.motionY + this.motionZ * this.motionZ);
+					int var20 = (int)Math.ceil((double)var19 * 2.0D);
+					if(this.field_35140_d) {
+						var20 = var20 * 3 / 2 + 1;
+					}
+
+					if(var3.entityHit.attackEntityFrom(this.owner, var20)) {
+						if(var3.entityHit instanceof EntityLiving) {
+							++((EntityLiving)var3.entityHit).field_35172_bP;
+						}
+
 						this.worldObj.playSoundAtEntity(this, "random.drr", 1.0F, 1.2F / (this.rand.nextFloat() * 0.2F + 0.9F));
 						this.setEntityDead();
 					} else {
@@ -187,6 +199,13 @@
 					this.worldObj.playSoundAtEntity(this, "random.drr", 1.0F, 1.2F / (this.rand.nextFloat() * 0.2F + 0.9F));
 					this.inGround = true;
 					this.arrowShake = 7;
+					this.field_35140_d = false;
+				}
+			}
+
+			if(this.field_35140_d) {
+				for(var8 = 0; var8 < 4; ++var8) {
+					this.worldObj.spawnParticle("crit", this.posX + this.motionX * (double)var8 / 4.0D, this.posY + this.motionY * (double)var8 / 4.0D, this.posZ + this.motionZ * (double)var8 / 4.0D, -this.motionX, -this.motionY + 0.2D, -this.motionZ);
 				}
 			}
 
--- net/minecraft/src/ModelSheep2.java
+++ net/minecraft/src/ModelSheep2.java
@@ -3,10 +3,10 @@
 public class ModelSheep2 extends ModelQuadruped {
 	public ModelSheep2() {
 		super(12, 0.0F);
-		this.head = new ModelRenderer(0, 0);
+		this.head = new ModelRenderer(this, 0, 0);
 		this.head.addBox(-3.0F, -4.0F, -6.0F, 6, 6, 8, 0.0F);
 		this.head.setRotationPoint(0.0F, 6.0F, -8.0F);
-		this.body = new ModelRenderer(28, 8);
+		this.body = new ModelRenderer(this, 28, 8);
 		this.body.addBox(-4.0F, -10.0F, -7.0F, 8, 16, 6, 0.0F);
 		this.body.setRotationPoint(0.0F, 5.0F, 2.0F);
 	}
--- net/minecraft/src/RenderPlayer.java
+++ net/minecraft/src/RenderPlayer.java
@@ -39,6 +39,13 @@
 	public void renderPlayer(EntityPlayer var1, double var2, double var4, double var6, float var8, float var9) {
 		ItemStack var10 = var1.inventory.getCurrentItem();
 		this.modelArmorChestplate.field_1278_i = this.modelArmor.field_1278_i = this.modelBipedMain.field_1278_i = var10 != null;
+		if(var10 != null && var1.getItemInUseCount() > 0) {
+			EnumAction var11 = var10.getItemUseAction();
+			if(var11 == EnumAction.bow) {
+				this.modelArmorChestplate.field_40333_u = this.modelArmor.field_40333_u = this.modelBipedMain.field_40333_u = true;
+			}
+		}
+
 		this.modelArmorChestplate.isSneak = this.modelArmor.isSneak = this.modelBipedMain.isSneak = var1.isSneaking();
 		double var11 = var4 - (double)var1.yOffset;
 		if(var1.isSneaking() && !(var1 instanceof EntityPlayerSP)) {
@@ -46,6 +53,7 @@
 		}
 
 		super.doRenderLiving(var1, var2, var11, var6, var8, var9);
+		this.modelArmorChestplate.field_40333_u = this.modelArmor.field_40333_u = this.modelBipedMain.field_40333_u = false;
 		this.modelArmorChestplate.isSneak = this.modelArmor.isSneak = this.modelBipedMain.isSneak = false;
 		this.modelArmorChestplate.field_1278_i = this.modelArmor.field_1278_i = this.modelBipedMain.field_1278_i = false;
 	}
@@ -101,6 +109,7 @@
 	}
 
 	protected void renderSpecials(EntityPlayer var1, float var2) {
+		super.renderEquippedItems(var1, var2);
 		ItemStack var3 = var1.inventory.armorItemInSlot(3);
 		if(var3 != null && var3.getItem().shiftedIndex < 256) {
 			GL11.glPushMatrix();
@@ -189,6 +198,13 @@
 				GL11.glRotatef(20.0F, 1.0F, 0.0F, 0.0F);
 				GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
 				GL11.glScalef(var5, -var5, var5);
+			} else if(var21.itemID == Item.bow.shiftedIndex) {
+				var5 = 10.0F / 16.0F;
+				GL11.glTranslatef(0.0F, 2.0F / 16.0F, 5.0F / 16.0F);
+				GL11.glRotatef(-20.0F, 0.0F, 1.0F, 0.0F);
+				GL11.glScalef(var5, -var5, var5);
+				GL11.glRotatef(-100.0F, 1.0F, 0.0F, 0.0F);
+				GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
 			} else if(Item.itemsList[var21.itemID].isFull3D()) {
 				var5 = 10.0F / 16.0F;
 				if(Item.itemsList[var21.itemID].shouldRotateAroundWhenRendering()) {
--- net/minecraft/src/GuiIngame.java
+++ net/minecraft/src/GuiIngame.java
@@ -352,7 +352,7 @@
 				GL11.glTranslatef((float)(-(var2 + 8)), (float)(-(var3 + 12)), 0.0F);
 			}
 
-			itemRenderer.renderItemIntoGUI(this.mc.fontRenderer, this.mc.renderEngine, var5, var2, var3);
+			itemRenderer.renderItemIntoGUI(this.mc.fontRenderer, this.mc.renderEngine, var5, var2, var3, this.mc.thePlayer);
 			if(var6 > 0.0F) {
 				GL11.glPopMatrix();
 			}
--- net/minecraft/src/SignModel.java
+++ net/minecraft/src/SignModel.java
@@ -1,12 +1,12 @@
 package net.minecraft.src;
 
-public class SignModel {
-	public ModelRenderer signBoard = new ModelRenderer(0, 0);
+public class SignModel extends ModelBase {
+	public ModelRenderer signBoard = new ModelRenderer(this, 0, 0);
 	public ModelRenderer signStick;
 
 	public SignModel() {
 		this.signBoard.addBox(-12.0F, -14.0F, -1.0F, 24, 12, 2, 0.0F);
-		this.signStick = new ModelRenderer(0, 14);
+		this.signStick = new ModelRenderer(this, 0, 14);
 		this.signStick.addBox(-1.0F, -2.0F, -1.0F, 2, 14, 2, 0.0F);
 	}
 
--- net/minecraft/src/GuiContainer.java
+++ net/minecraft/src/GuiContainer.java
@@ -54,7 +54,7 @@
 		InventoryPlayer var12 = this.mc.thePlayer.inventory;
 		if(var12.getItemStack() != null) {
 			GL11.glTranslatef(0.0F, 0.0F, 32.0F);
-			itemRenderer.renderItemIntoGUI(this.fontRenderer, this.mc.renderEngine, var12.getItemStack(), var1 - var4 - 8, var2 - var5 - 8);
+			itemRenderer.renderItemIntoGUI(this.fontRenderer, this.mc.renderEngine, var12.getItemStack(), var1 - var4 - 8, var2 - var5 - 8, this.mc.thePlayer);
 			itemRenderer.renderItemOverlayIntoGUI(this.fontRenderer, this.mc.renderEngine, var12.getItemStack(), var1 - var4 - 8, var2 - var5 - 8);
 		}
 
@@ -100,7 +100,7 @@
 			}
 		}
 
-		itemRenderer.renderItemIntoGUI(this.fontRenderer, this.mc.renderEngine, var4, var2, var3);
+		itemRenderer.renderItemIntoGUI(this.fontRenderer, this.mc.renderEngine, var4, var2, var3, this.mc.thePlayer);
 		itemRenderer.renderItemOverlayIntoGUI(this.fontRenderer, this.mc.renderEngine, var4, var2, var3);
 	}
 
--- net/minecraft/src/ModelCow.java
+++ net/minecraft/src/ModelCow.java
@@ -7,20 +7,20 @@
 
 	public ModelCow() {
 		super(12, 0.0F);
-		this.head = new ModelRenderer(0, 0);
+		this.head = new ModelRenderer(this, 0, 0);
 		this.head.addBox(-4.0F, -4.0F, -6.0F, 8, 8, 6, 0.0F);
 		this.head.setRotationPoint(0.0F, 4.0F, -8.0F);
-		this.horn1 = new ModelRenderer(22, 0);
+		this.horn1 = new ModelRenderer(this, 22, 0);
 		this.horn1.addBox(-4.0F, -5.0F, -4.0F, 1, 3, 1, 0.0F);
 		this.horn1.setRotationPoint(0.0F, 3.0F, -7.0F);
-		this.horn2 = new ModelRenderer(22, 0);
+		this.horn2 = new ModelRenderer(this, 22, 0);
 		this.horn2.addBox(3.0F, -5.0F, -4.0F, 1, 3, 1, 0.0F);
 		this.horn2.setRotationPoint(0.0F, 3.0F, -7.0F);
-		this.udders = new ModelRenderer(52, 0);
+		this.udders = new ModelRenderer(this, 52, 0);
 		this.udders.addBox(-2.0F, -3.0F, 0.0F, 4, 6, 2, 0.0F);
 		this.udders.setRotationPoint(0.0F, 14.0F, 6.0F);
 		this.udders.rotateAngleX = (float)Math.PI * 0.5F;
-		this.body = new ModelRenderer(18, 4);
+		this.body = new ModelRenderer(this, 18, 4);
 		this.body.addBox(-6.0F, -10.0F, -7.0F, 12, 18, 10, 0.0F);
 		this.body.setRotationPoint(0.0F, 5.0F, 2.0F);
 		--this.leg1.rotationPointX;
--- net/minecraft/src/ItemSoup.java
+++ net/minecraft/src/ItemSoup.java
@@ -5,7 +5,7 @@
 		super(var1, var2, false);
 	}
 
-	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
+	public ItemStack onFoodEaten(ItemStack var1, World var2, EntityPlayer var3) {
 		super.onItemRightClick(var1, var2, var3);
 		return new ItemStack(Item.bowlEmpty);
 	}
--- net/minecraft/src/RenderLiving.java
+++ net/minecraft/src/RenderLiving.java
@@ -1,5 +1,6 @@
 package net.minecraft.src;
 
+import java.util.Random;
 import net.minecraft.client.Minecraft;
 import org.lwjgl.opengl.GL11;
 import org.lwjgl.opengl.GL12;
@@ -141,6 +142,38 @@
 	}
 
 	protected void renderEquippedItems(EntityLiving var1, float var2) {
+		if(var1.field_35172_bP > 0) {
+			EntityArrow var3 = new EntityArrow(var1.worldObj, var1.posX, var1.posY, var1.posZ);
+			Random var4 = new Random((long)var1.entityId);
+
+			for(int var5 = 0; var5 < var1.field_35172_bP; ++var5) {
+				GL11.glPushMatrix();
+				ModelRenderer var6 = this.mainModel.func_35393_a(var4);
+				var6.postRender(1.0F / 16.0F);
+				float var7 = var4.nextFloat();
+				float var8 = var4.nextFloat();
+				float var9 = var4.nextFloat();
+				float var10 = (var6.field_35977_i + (var6.field_35973_l - var6.field_35977_i) * var7) / 16.0F;
+				float var11 = (var6.field_35975_j + (var6.field_35974_m - var6.field_35975_j) * var8) / 16.0F;
+				float var12 = (var6.field_35976_k + (var6.field_35972_n - var6.field_35976_k) * var9) / 16.0F;
+				GL11.glTranslatef(var10, var11, var12);
+				var7 = var7 * 2.0F - 1.0F;
+				var8 = var8 * 2.0F - 1.0F;
+				var9 = var9 * 2.0F - 1.0F;
+				var7 *= -1.0F;
+				var8 *= -1.0F;
+				var9 *= -1.0F;
+				float var13 = MathHelper.sqrt_float(var7 * var7 + var9 * var9);
+				var3.prevRotationYaw = var3.rotationYaw = (float)(Math.atan2((double)var7, (double)var9) * 180.0D / (double)((float)Math.PI));
+				var3.prevRotationPitch = var3.rotationPitch = (float)(Math.atan2((double)var8, (double)var13) * 180.0D / (double)((float)Math.PI));
+				double var14 = 0.0D;
+				double var16 = 0.0D;
+				double var18 = 0.0D;
+				float var20 = 0.0F;
+				this.renderManager.renderEntityWithPosYaw(var3, var14, var16, var18, var20, var2);
+				GL11.glPopMatrix();
+			}
+		}
 	}
 
 	protected boolean func_27005_b(EntityLiving var1, int var2, float var3) {
--- net/minecraft/src/ModelSquid.java
+++ net/minecraft/src/ModelSquid.java
@@ -6,12 +6,12 @@
 
 	public ModelSquid() {
 		byte var1 = -16;
-		this.squidBody = new ModelRenderer(0, 0);
+		this.squidBody = new ModelRenderer(this, 0, 0);
 		this.squidBody.addBox(-6.0F, -8.0F, -6.0F, 12, 16, 12);
 		this.squidBody.rotationPointY += (float)(24 + var1);
 
 		for(int var2 = 0; var2 < this.squidTentacles.length; ++var2) {
-			this.squidTentacles[var2] = new ModelRenderer(48, 0);
+			this.squidTentacles[var2] = new ModelRenderer(this, 48, 0);
 			double var3 = (double)var2 * Math.PI * 2.0D / (double)this.squidTentacles.length;
 			float var5 = (float)Math.cos(var3) * 5.0F;
 			float var6 = (float)Math.sin(var3) * 5.0F;
--- net/minecraft/client/Minecraft.java
+++ net/minecraft/client/Minecraft.java
@@ -1006,6 +1006,10 @@
 														this.clickMouse(1);
 														this.mouseTicksRan = this.ticksRan;
 													}
+
+													if(!Mouse.isButtonDown(1)) {
+														this.playerController.onStoppedUsingItem(this.thePlayer);
+													}
 												}
 
 												this.func_6254_a(0, this.currentScreen == null && Mouse.isButtonDown(0) && this.inGameHasFocus);
--- net/minecraft/src/ModelQuadruped.java
+++ net/minecraft/src/ModelQuadruped.java
@@ -1,7 +1,7 @@
 package net.minecraft.src;
 
 public class ModelQuadruped extends ModelBase {
-	public ModelRenderer head = new ModelRenderer(0, 0);
+	public ModelRenderer head = new ModelRenderer(this, 0, 0);
 	public ModelRenderer body;
 	public ModelRenderer leg1;
 	public ModelRenderer leg2;
@@ -11,19 +11,19 @@
 	public ModelQuadruped(int var1, float var2) {
 		this.head.addBox(-4.0F, -4.0F, -8.0F, 8, 8, 8, var2);
 		this.head.setRotationPoint(0.0F, (float)(18 - var1), -6.0F);
-		this.body = new ModelRenderer(28, 8);
+		this.body = new ModelRenderer(this, 28, 8);
 		this.body.addBox(-5.0F, -10.0F, -7.0F, 10, 16, 8, var2);
 		this.body.setRotationPoint(0.0F, (float)(17 - var1), 2.0F);
-		this.leg1 = new ModelRenderer(0, 16);
+		this.leg1 = new ModelRenderer(this, 0, 16);
 		this.leg1.addBox(-2.0F, 0.0F, -2.0F, 4, var1, 4, var2);
 		this.leg1.setRotationPoint(-3.0F, (float)(24 - var1), 7.0F);
-		this.leg2 = new ModelRenderer(0, 16);
+		this.leg2 = new ModelRenderer(this, 0, 16);
 		this.leg2.addBox(-2.0F, 0.0F, -2.0F, 4, var1, 4, var2);
 		this.leg2.setRotationPoint(3.0F, (float)(24 - var1), 7.0F);
-		this.leg3 = new ModelRenderer(0, 16);
+		this.leg3 = new ModelRenderer(this, 0, 16);
 		this.leg3.addBox(-2.0F, 0.0F, -2.0F, 4, var1, 4, var2);
 		this.leg3.setRotationPoint(-3.0F, (float)(24 - var1), -5.0F);
-		this.leg4 = new ModelRenderer(0, 16);
+		this.leg4 = new ModelRenderer(this, 0, 16);
 		this.leg4.addBox(-2.0F, 0.0F, -2.0F, 4, var1, 4, var2);
 		this.leg4.setRotationPoint(3.0F, (float)(24 - var1), -5.0F);
 	}
--- net/minecraft/src/ModelCreeper.java
+++ net/minecraft/src/ModelCreeper.java
@@ -15,25 +15,25 @@
 
 	public ModelCreeper(float var1) {
 		byte var2 = 4;
-		this.head = new ModelRenderer(0, 0);
+		this.head = new ModelRenderer(this, 0, 0);
 		this.head.addBox(-4.0F, -8.0F, -4.0F, 8, 8, 8, var1);
 		this.head.setRotationPoint(0.0F, (float)var2, 0.0F);
-		this.field_1270_b = new ModelRenderer(32, 0);
+		this.field_1270_b = new ModelRenderer(this, 32, 0);
 		this.field_1270_b.addBox(-4.0F, -8.0F, -4.0F, 8, 8, 8, var1 + 0.5F);
 		this.field_1270_b.setRotationPoint(0.0F, (float)var2, 0.0F);
-		this.body = new ModelRenderer(16, 16);
+		this.body = new ModelRenderer(this, 16, 16);
 		this.body.addBox(-4.0F, 0.0F, -2.0F, 8, 12, 4, var1);
 		this.body.setRotationPoint(0.0F, (float)var2, 0.0F);
-		this.leg1 = new ModelRenderer(0, 16);
+		this.leg1 = new ModelRenderer(this, 0, 16);
 		this.leg1.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg1.setRotationPoint(-2.0F, (float)(12 + var2), 4.0F);
-		this.leg2 = new ModelRenderer(0, 16);
+		this.leg2 = new ModelRenderer(this, 0, 16);
 		this.leg2.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg2.setRotationPoint(2.0F, (float)(12 + var2), 4.0F);
-		this.leg3 = new ModelRenderer(0, 16);
+		this.leg3 = new ModelRenderer(this, 0, 16);
 		this.leg3.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg3.setRotationPoint(-2.0F, (float)(12 + var2), -4.0F);
-		this.leg4 = new ModelRenderer(0, 16);
+		this.leg4 = new ModelRenderer(this, 0, 16);
 		this.leg4.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg4.setRotationPoint(2.0F, (float)(12 + var2), -4.0F);
 	}
--- /dev/null
+++ net/minecraft/src/EnumAction.java
@@ -1,0 +1,7 @@
+package net.minecraft.src;
+
+public enum EnumAction {
+	none,
+	eat,
+	bow;
+}
--- net/minecraft/src/GuiOptions.java
+++ net/minecraft/src/GuiOptions.java
@@ -4,7 +4,7 @@
 	private GuiScreen parentScreen;
 	protected String screenTitle = "Options";
 	private GameSettings options;
-	private static EnumOptions[] field_22135_k = new EnumOptions[]{EnumOptions.MUSIC, EnumOptions.SOUND, EnumOptions.INVERT_MOUSE, EnumOptions.SENSITIVITY, EnumOptions.DIFFICULTY};
+	private static EnumOptions[] field_22135_k = new EnumOptions[]{EnumOptions.MUSIC, EnumOptions.SOUND, EnumOptions.INVERT_MOUSE, EnumOptions.SENSITIVITY, EnumOptions.FOV, EnumOptions.DIFFICULTY};
 
 	public GuiOptions(GuiScreen var1, GameSettings var2) {
 		this.parentScreen = var1;
--- net/minecraft/src/InventoryPlayer.java
+++ net/minecraft/src/InventoryPlayer.java
@@ -128,6 +128,11 @@
 		}
 	}
 
+	public boolean hasItem(int var1) {
+		int var2 = this.getInventorySlotContainItem(var1);
+		return var2 >= 0;
+	}
+
 	public boolean addItemStackToInventory(ItemStack var1) {
 		int var2;
 		if(var1.isItemDamaged()) {
--- net/minecraft/src/EntityPlayerSP.java
+++ net/minecraft/src/EntityPlayerSP.java
@@ -74,8 +74,13 @@
 		if(this.timeUntilPortal > 0) {
 			--this.timeUntilPortal;
 		}
-
+		
 		this.movementInput.updatePlayerMoveState(this);
+		if(this.isUsingItem()) {
+			this.movementInput.moveStrafe *= 0.2F;
+			this.movementInput.moveForward *= 0.2F;
+		}
+		
 		if(this.movementInput.sneak && this.ySize < 0.2F) {
 			this.ySize = 0.2F;
 		}
@@ -87,6 +92,24 @@
 		super.onLivingUpdate();
 	}
 
+	public float getFOVMultiplier() {
+		float var1 = 1.0F;
+		var1 *= (0.1F * 1.0F / 0.1F + 1.0F) / 2.0F;
+		if(this.isUsingItem() && this.getItemInUse().itemID == Item.bow.shiftedIndex) {
+			int var2 = this.getItemInUseDuration();
+			float var3 = (float)var2 / 20.0F;
+			if(var3 > 1.0F) {
+				var3 = 1.0F;
+			} else {
+				var3 *= var3;
+			}
+
+			var1 *= 1.0F - var3 * 0.15F;
+		}
+
+		return var1;
+	}
+
 	public void resetPlayerKeyState() {
 		this.movementInput.resetKeyState();
 	}
@@ -94,7 +117,7 @@
 	public void handleKeyPress(int var1, boolean var2) {
 		this.movementInput.checkKeyForMovementInput(var1, var2);
 	}
-
+	
 	public void writeEntityToNBT(NBTTagCompound var1) {
 		super.writeEntityToNBT(var1);
 		var1.setInteger("Score", this.score);
--- net/minecraft/src/GuiAchievement.java
+++ net/minecraft/src/GuiAchievement.java
@@ -114,7 +114,7 @@
 				GL11.glEnable(GL12.GL_RESCALE_NORMAL);
 				GL11.glEnable(GL11.GL_COLOR_MATERIAL);
 				GL11.glEnable(GL11.GL_LIGHTING);
-				this.itemRender.renderItemIntoGUI(this.theGame.fontRenderer, this.theGame.renderEngine, this.theAchievement.theItemStack, var5 + 8, var6 + 8);
+				this.itemRender.renderItemIntoGUI(this.theGame.fontRenderer, this.theGame.renderEngine, this.theAchievement.theItemStack, var5 + 8, var6 + 8, null);
 				GL11.glDisable(GL11.GL_LIGHTING);
 				GL11.glDepthMask(true);
 				GL11.glEnable(GL11.GL_DEPTH_TEST);
--- net/minecraft/src/EntityPlayer.java
+++ net/minecraft/src/EntityPlayer.java
@@ -34,6 +34,8 @@
 	protected boolean inPortal = false;
 	public float timeInPortal;
 	public float prevTimeInPortal;
+	private ItemStack itemInUse;
+	public int itemInUseCount;
 	private int damageRemainder = 0;
 	public EntityFish fishEntity = null;
 
@@ -56,7 +58,67 @@
 		this.dataWatcher.addObject(16, Byte.valueOf((byte)0));
 	}
 
+	public ItemStack getItemInUse() {
+		return this.itemInUse;
+	}
+
+	public int getItemInUseCount() {
+		return this.itemInUseCount;
+	}
+
+	public boolean isUsingItem() {
+		return this.itemInUse != null;
+	}
+	
+	public int getItemInUseDuration() {
+		return this.isUsingItem() ? this.itemInUse.getMaxItemUseDuration() - this.itemInUseCount : 0;
+	}
+
+	public void stopUsingItem() {
+		if(this.itemInUse != null) {
+			this.itemInUse.onPlayerStoppedUsing(this.worldObj, this, this.itemInUseCount);
+		}
+
+		this.clearItemInUse();
+	}
+
+	public void clearItemInUse() {
+		this.itemInUse = null;
+		this.itemInUseCount = 0;
+		if(!this.worldObj.multiplayerWorld) {
+			this.setEating(false);
+		}
+
+	}
+
 	public void onUpdate() {
+		if(this.itemInUse != null) {
+			ItemStack var1 = this.inventory.getCurrentItem();
+			if(var1 != this.itemInUse) {
+				this.clearItemInUse();
+			} else {
+				if(this.itemInUseCount <= 25 && this.itemInUseCount % 4 == 0) {
+					this.func_35201_a(var1, 5);
+				}
+
+				if(--this.itemInUseCount == 0 && !this.worldObj.multiplayerWorld) {
+					if(this.itemInUse != null) {
+						this.func_35201_a(this.itemInUse, 16);
+						int var2 = this.itemInUse.stackSize;
+						ItemStack var3 = this.itemInUse.onFoodEaten(this.worldObj, this);
+						if(var3 != this.itemInUse || var3 != null && var3.stackSize != var2) {
+							this.inventory.mainInventory[this.inventory.currentItem] = var3;
+							if(var3.stackSize == 0) {
+								this.inventory.mainInventory[this.inventory.currentItem] = null;
+							}
+						}
+
+						this.clearItemInUse();
+					}
+				}
+			}
+		}
+
 		if(this.isPlayerSleeping()) {
 			++this.sleepTimer;
 			if(this.sleepTimer > 100) {
@@ -124,6 +186,24 @@
 
 	}
 
+	protected void func_35201_a(ItemStack var1, int var2) {
+		if(var1.getItemUseAction() == EnumAction.eat) {
+			for(int var3 = 0; var3 < var2; ++var3) {
+				Vec3D var4 = Vec3D.createVector(((double)this.rand.nextFloat() - 0.5D) * 0.1D, Math.random() * 0.1D + 0.1D, 0.0D);
+				var4.rotateAroundX(-this.rotationPitch * (float)Math.PI / 180.0F);
+				var4.rotateAroundY(-this.rotationYaw * (float)Math.PI / 180.0F);
+				Vec3D var5 = Vec3D.createVector(((double)this.rand.nextFloat() - 0.5D) * 0.3D, (double)(-this.rand.nextFloat()) * 0.6D - 0.3D, 0.6D);
+				var5.rotateAroundX(-this.rotationPitch * (float)Math.PI / 180.0F);
+				var5.rotateAroundY(-this.rotationYaw * (float)Math.PI / 180.0F);
+				var5 = var5.addVector(this.posX, this.posY + (double)this.getEyeHeight(), this.posZ);
+				this.worldObj.spawnParticle("iconcrack_" + var1.getItem().shiftedIndex, var5.xCoord, var5.yCoord, var5.zCoord, var4.xCoord, var4.yCoord + 0.05D, var4.zCoord);
+			}
+
+			this.worldObj.playSoundAtEntity(this, "random.eat", 0.5F + 0.5F * (float)this.rand.nextInt(2), (this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F + 1.0F);
+		}
+
+	}
+
 	protected boolean isMovementBlocked() {
 		return this.health <= 0 || this.isPlayerSleeping();
 	}
@@ -796,12 +876,27 @@
 	}
 
 	public int getItemIcon(ItemStack var1) {
-		int var2 = super.getItemIcon(var1);
+		int var3 = super.getItemIcon(var1);
 		if(var1.itemID == Item.fishingRod.shiftedIndex && this.fishEntity != null) {
-			var2 = var1.getIconIndex() + 16;
+			var3 = var1.getIconIndex() + 16;
+		} else {
+			if(this.itemInUse != null && var1.itemID == Item.bow.shiftedIndex) {
+				int var4 = var1.getMaxItemUseDuration() - this.itemInUseCount;
+				if(var4 >= 18) {
+					return 133;
+				}
+
+				if(var4 > 13) {
+					return 117;
+				}
+
+				if(var4 > 0) {
+					return 101;
+				}
+			}
 		}
 
-		return var2;
+		return var3;
 	}
 
 	public void setInPortal() {
@@ -809,6 +904,17 @@
 			this.timeUntilPortal = 10;
 		} else {
 			this.inPortal = true;
+		}
+	}
+
+	public void setItemInUse(ItemStack var1, int var2) {
+		if(var1 != this.itemInUse) {
+			this.itemInUse = var1;
+			this.itemInUseCount = var2;
+			if(!this.worldObj.multiplayerWorld) {
+				this.setEating(true);
+			}
+
 		}
 	}
 }
--- net/minecraft/src/RenderGlobal.java
+++ net/minecraft/src/RenderGlobal.java
@@ -1310,6 +1310,12 @@
 					this.mc.effectRenderer.addEffect(new EntitySlimeFX(this.worldObj, var2, var4, var6, Item.slimeBall));
 				} else if(var1.equals("heart")) {
 					this.mc.effectRenderer.addEffect(new EntityHeartFX(this.worldObj, var2, var4, var6, var8, var10, var12));
+				} else {
+					int var24;
+					if(var1.startsWith("iconcrack_")) {
+						var24 = Integer.parseInt(var1.substring(var1.indexOf("_") + 1));
+						this.mc.effectRenderer.addEffect(new EntityBreakingFX(this.worldObj, var2, var4, var6, var8, var10, var12, Item.itemsList[var24]));
+					}
 				}
 
 			}
--- net/minecraft/src/EntitySkeleton.java
+++ net/minecraft/src/EntitySkeleton.java
@@ -36,7 +36,7 @@
 			double var3 = var1.posX - this.posX;
 			double var5 = var1.posZ - this.posZ;
 			if(this.attackTime == 0) {
-				EntityArrow var7 = new EntityArrow(this.worldObj, this);
+				EntityArrow var7 = new EntityArrow(this.worldObj, this, 1.0F);
 				var7.posY += (double)1.4F;
 				double var8 = var1.posY + (double)var1.getEyeHeight() - (double)0.2F - var7.posY;
 				float var10 = MathHelper.sqrt_double(var3 * var3 + var5 * var5) * 0.2F;
--- net/minecraft/src/GuiAchievements.java
+++ net/minecraft/src/GuiAchievements.java
@@ -274,7 +274,7 @@
 
 				GL11.glEnable(GL11.GL_LIGHTING);
 				GL11.glEnable(GL11.GL_CULL_FACE);
-				var29.renderItemIntoGUI(this.mc.fontRenderer, this.mc.renderEngine, var30.theItemStack, var33 + 3, var34 + 3);
+				var29.renderItemIntoGUI(this.mc.fontRenderer, this.mc.renderEngine, var30.theItemStack, var33 + 3, var34 + 3, null);
 				GL11.glDisable(GL11.GL_LIGHTING);
 				if(!this.field_27120_x.func_27181_b(var30)) {
 					var29.field_27004_a = true;
--- net/minecraft/src/ItemStack.java
+++ net/minecraft/src/ItemStack.java
@@ -72,6 +72,10 @@
 		return this.getItem().onItemRightClick(this, var1, var2);
 	}
 
+	public ItemStack onFoodEaten(World var1, EntityPlayer var2) {
+		return this.getItem().onFoodEaten(this, var1, var2);
+	}
+
 	public NBTTagCompound writeToNBT(NBTTagCompound var1) {
 		var1.setShort("id", (short)this.itemID);
 		var1.setByte("Count", (byte)this.stackSize);
@@ -214,5 +218,17 @@
 
 	public boolean isStackEqual(ItemStack var1) {
 		return this.itemID == var1.itemID && this.stackSize == var1.stackSize && this.itemDamage == var1.itemDamage;
+	}
+
+	public int getMaxItemUseDuration() {
+		return this.getItem().getMaxItemUseDuration(this);
+	}
+
+	public EnumAction getItemUseAction() {
+		return this.getItem().getItemUseAction(this);
+	}
+
+	public void onPlayerStoppedUsing(World var1, EntityPlayer var2, int var3) {
+		this.getItem().onPlayerStoppedUsing(this, var1, var2, var3);
 	}
 }
--- net/minecraft/src/Entity.java
+++ net/minecraft/src/Entity.java
@@ -1027,6 +1027,14 @@
 		return this.getEntityFlag(1);
 	}
 
+	public boolean isEating() {
+		return this.getEntityFlag(4);
+	}
+
+	public void setEating(boolean var1) {
+		this.setEntityFlag(4, var1);
+	}
+
 	protected boolean getEntityFlag(int var1) {
 		return (this.dataWatcher.getWatchableObjectByte(0) & 1 << var1) != 0;
 	}
--- net/minecraft/src/ModelSlime.java
+++ net/minecraft/src/ModelSlime.java
@@ -7,16 +7,16 @@
 	ModelRenderer slimeMouth;
 
 	public ModelSlime(int var1) {
-		this.slimeBodies = new ModelRenderer(0, var1);
+		this.slimeBodies = new ModelRenderer(this, 0, var1);
 		this.slimeBodies.addBox(-4.0F, 16.0F, -4.0F, 8, 8, 8);
 		if(var1 > 0) {
-			this.slimeBodies = new ModelRenderer(0, var1);
+			this.slimeBodies = new ModelRenderer(this, 0, var1);
 			this.slimeBodies.addBox(-3.0F, 17.0F, -3.0F, 6, 6, 6);
-			this.slimeRightEye = new ModelRenderer(32, 0);
+			this.slimeRightEye = new ModelRenderer(this, 32, 0);
 			this.slimeRightEye.addBox(-3.25F, 18.0F, -3.5F, 2, 2, 2);
-			this.slimeLeftEye = new ModelRenderer(32, 4);
+			this.slimeLeftEye = new ModelRenderer(this, 32, 4);
 			this.slimeLeftEye.addBox(1.25F, 18.0F, -3.5F, 2, 2, 2);
-			this.slimeMouth = new ModelRenderer(32, 8);
+			this.slimeMouth = new ModelRenderer(this, 32, 8);
 			this.slimeMouth.addBox(0.0F, 21.0F, -3.5F, 1, 1, 1);
 		}
 
--- net/minecraft/src/ModelMinecart.java
+++ net/minecraft/src/ModelMinecart.java
@@ -4,12 +4,12 @@
 	public ModelRenderer[] sideModels = new ModelRenderer[7];
 
 	public ModelMinecart() {
-		this.sideModels[0] = new ModelRenderer(0, 10);
-		this.sideModels[1] = new ModelRenderer(0, 0);
-		this.sideModels[2] = new ModelRenderer(0, 0);
-		this.sideModels[3] = new ModelRenderer(0, 0);
-		this.sideModels[4] = new ModelRenderer(0, 0);
-		this.sideModels[5] = new ModelRenderer(44, 10);
+		this.sideModels[0] = new ModelRenderer(this, 0, 10);
+		this.sideModels[1] = new ModelRenderer(this, 0, 0);
+		this.sideModels[2] = new ModelRenderer(this, 0, 0);
+		this.sideModels[3] = new ModelRenderer(this, 0, 0);
+		this.sideModels[4] = new ModelRenderer(this, 0, 0);
+		this.sideModels[5] = new ModelRenderer(this, 44, 10);
 		byte var1 = 20;
 		byte var2 = 8;
 		byte var3 = 16;
--- net/minecraft/src/ItemBow.java
+++ net/minecraft/src/ItemBow.java
@@ -4,14 +4,52 @@
 	public ItemBow(int var1) {
 		super(var1);
 		this.maxStackSize = 1;
+		this.setMaxDamage(384);
 	}
 
-	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
-		if(var3.inventory.consumeInventoryItem(Item.arrow.shiftedIndex)) {
-			var2.playSoundAtEntity(var3, "random.bow", 1.0F, 1.0F / (itemRand.nextFloat() * 0.4F + 0.8F));
+	public void onPlayerStoppedUsing(ItemStack var1, World var2, EntityPlayer var3, int var4) {
+		if(var3.inventory.hasItem(Item.arrow.shiftedIndex)) {
+			int var5 = this.getMaxItemUseDuration(var1) - var4;
+			float var6 = (float)var5 / 20.0F;
+			var6 = (var6 * var6 + var6 * 2.0F) / 3.0F;
+			if((double)var6 < 0.1D) {
+				return;
+			}
+
+			if(var6 > 1.0F) {
+				var6 = 1.0F;
+			}
+
+			EntityArrow var7 = new EntityArrow(var2, var3, var6 * 2.0F);
+			if(var6 == 1.0F) {
+				var7.field_35140_d = true;
+			}
+
+			var1.damageItem(1, var3);
+			var2.playSoundAtEntity(var3, "random.bow", 1.0F, 1.0F / (itemRand.nextFloat() * 0.4F + 1.2F) + var6 * 0.5F);
+			var3.inventory.consumeInventoryItem(Item.arrow.shiftedIndex);
 			if(!var2.multiplayerWorld) {
-				var2.entityJoinedWorld(new EntityArrow(var2, var3));
+				var2.entityJoinedWorld(var7);
 			}
+		}
+
+	}
+
+	public ItemStack onFoodEaten(ItemStack var1, World var2, EntityPlayer var3) {
+		return var1;
+	}
+
+	public int getMaxItemUseDuration(ItemStack var1) {
+		return 72000;
+	}
+
+	public EnumAction getItemUseAction(ItemStack var1) {
+		return EnumAction.bow;
+	}
+
+	public ItemStack onItemRightClick(ItemStack var1, World var2, EntityPlayer var3) {
+		if(var3.inventory.hasItem(Item.arrow.shiftedIndex)) {
+			var3.setItemInUse(var1, this.getMaxItemUseDuration(var1));
 		}
 
 		return var1;
--- net/minecraft/src/EnumOptions.java
+++ net/minecraft/src/EnumOptions.java
@@ -5,6 +5,7 @@
 	SOUND("options.sound", true, false),
 	INVERT_MOUSE("options.invertMouse", false, true),
 	SENSITIVITY("options.sensitivity", true, false),
+	FOV("options.fov", true, false),
 	RENDER_DISTANCE("options.renderDistance", false, false),
 	VIEW_BOBBING("options.viewBobbing", false, true),
 	ANAGLYPH("options.anaglyph", false, true),
--- net/minecraft/src/ModelSpider.java
+++ net/minecraft/src/ModelSpider.java
@@ -16,37 +16,37 @@
 	public ModelSpider() {
 		float var1 = 0.0F;
 		byte var2 = 15;
-		this.spiderHead = new ModelRenderer(32, 4);
+		this.spiderHead = new ModelRenderer(this, 32, 4);
 		this.spiderHead.addBox(-4.0F, -4.0F, -8.0F, 8, 8, 8, var1);
 		this.spiderHead.setRotationPoint(0.0F, (float)(0 + var2), -3.0F);
-		this.spiderNeck = new ModelRenderer(0, 0);
+		this.spiderNeck = new ModelRenderer(this, 0, 0);
 		this.spiderNeck.addBox(-3.0F, -3.0F, -3.0F, 6, 6, 6, var1);
 		this.spiderNeck.setRotationPoint(0.0F, (float)var2, 0.0F);
-		this.spiderBody = new ModelRenderer(0, 12);
+		this.spiderBody = new ModelRenderer(this, 0, 12);
 		this.spiderBody.addBox(-5.0F, -4.0F, -6.0F, 10, 8, 12, var1);
 		this.spiderBody.setRotationPoint(0.0F, (float)(0 + var2), 9.0F);
-		this.spiderLeg1 = new ModelRenderer(18, 0);
+		this.spiderLeg1 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg1.addBox(-15.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg1.setRotationPoint(-4.0F, (float)(0 + var2), 2.0F);
-		this.spiderLeg2 = new ModelRenderer(18, 0);
+		this.spiderLeg2 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg2.addBox(-1.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg2.setRotationPoint(4.0F, (float)(0 + var2), 2.0F);
-		this.spiderLeg3 = new ModelRenderer(18, 0);
+		this.spiderLeg3 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg3.addBox(-15.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg3.setRotationPoint(-4.0F, (float)(0 + var2), 1.0F);
-		this.spiderLeg4 = new ModelRenderer(18, 0);
+		this.spiderLeg4 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg4.addBox(-1.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg4.setRotationPoint(4.0F, (float)(0 + var2), 1.0F);
-		this.spiderLeg5 = new ModelRenderer(18, 0);
+		this.spiderLeg5 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg5.addBox(-15.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg5.setRotationPoint(-4.0F, (float)(0 + var2), 0.0F);
-		this.spiderLeg6 = new ModelRenderer(18, 0);
+		this.spiderLeg6 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg6.addBox(-1.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg6.setRotationPoint(4.0F, (float)(0 + var2), 0.0F);
-		this.spiderLeg7 = new ModelRenderer(18, 0);
+		this.spiderLeg7 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg7.addBox(-15.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg7.setRotationPoint(-4.0F, (float)(0 + var2), -1.0F);
-		this.spiderLeg8 = new ModelRenderer(18, 0);
+		this.spiderLeg8 = new ModelRenderer(this, 18, 0);
 		this.spiderLeg8.addBox(-1.0F, -1.0F, -1.0F, 16, 2, 2, var1);
 		this.spiderLeg8.setRotationPoint(4.0F, (float)(0 + var2), -1.0F);
 	}
--- net/minecraft/src/ModelSkeleton.java
+++ net/minecraft/src/ModelSkeleton.java
@@ -3,17 +3,17 @@
 public class ModelSkeleton extends ModelZombie {
 	public ModelSkeleton() {
 		float var1 = 0.0F;
-		this.bipedRightArm = new ModelRenderer(40, 16);
+		this.bipedRightArm = new ModelRenderer(this, 40, 16);
 		this.bipedRightArm.addBox(-1.0F, -2.0F, -1.0F, 2, 12, 2, var1);
 		this.bipedRightArm.setRotationPoint(-5.0F, 2.0F, 0.0F);
-		this.bipedLeftArm = new ModelRenderer(40, 16);
+		this.bipedLeftArm = new ModelRenderer(this, 40, 16);
 		this.bipedLeftArm.mirror = true;
 		this.bipedLeftArm.addBox(-1.0F, -2.0F, -1.0F, 2, 12, 2, var1);
 		this.bipedLeftArm.setRotationPoint(5.0F, 2.0F, 0.0F);
-		this.bipedRightLeg = new ModelRenderer(0, 16);
+		this.bipedRightLeg = new ModelRenderer(this, 0, 16);
 		this.bipedRightLeg.addBox(-1.0F, 0.0F, -1.0F, 2, 12, 2, var1);
 		this.bipedRightLeg.setRotationPoint(-2.0F, 12.0F, 0.0F);
-		this.bipedLeftLeg = new ModelRenderer(0, 16);
+		this.bipedLeftLeg = new ModelRenderer(this, 0, 16);
 		this.bipedLeftLeg.mirror = true;
 		this.bipedLeftLeg.addBox(-1.0F, 0.0F, -1.0F, 2, 12, 2, var1);
 		this.bipedLeftLeg.setRotationPoint(2.0F, 12.0F, 0.0F);
--- net/minecraft/src/Item.java
+++ net/minecraft/src/Item.java
@@ -164,6 +164,10 @@
 		return var1;
 	}
 
+	public ItemStack onFoodEaten(ItemStack var1, World var2, EntityPlayer var3) {
+		return var1;
+	}
+
 	public int getItemStackLimit() {
 		return this.maxStackSize;
 	}
@@ -268,6 +272,17 @@
 	}
 
 	public void onCreated(ItemStack var1, World var2, EntityPlayer var3) {
+	}
+
+	public EnumAction getItemUseAction(ItemStack var1) {
+		return EnumAction.none;
+	}
+
+	public int getMaxItemUseDuration(ItemStack var1) {
+		return 0;
+	}
+
+	public void onPlayerStoppedUsing(ItemStack var1, World var2, EntityPlayer var3, int var4) {
 	}
 
 	static {
--- net/minecraft/src/ModelBiped.java
+++ net/minecraft/src/ModelBiped.java
@@ -13,6 +13,7 @@
 	public boolean field_1279_h;
 	public boolean field_1278_i;
 	public boolean isSneak;
+	public boolean field_40333_u;
 
 	public ModelBiped() {
 		this(0.0F);
@@ -26,30 +27,30 @@
 		this.field_1279_h = false;
 		this.field_1278_i = false;
 		this.isSneak = false;
-		this.bipedCloak = new ModelRenderer(0, 0);
+		this.bipedCloak = new ModelRenderer(this, 0, 0);
 		this.bipedCloak.addBox(-5.0F, 0.0F, -1.0F, 10, 16, 1, var1);
-		this.bipedEars = new ModelRenderer(24, 0);
+		this.bipedEars = new ModelRenderer(this, 24, 0);
 		this.bipedEars.addBox(-3.0F, -6.0F, -1.0F, 6, 6, 1, var1);
-		this.bipedHead = new ModelRenderer(0, 0);
+		this.bipedHead = new ModelRenderer(this, 0, 0);
 		this.bipedHead.addBox(-4.0F, -8.0F, -4.0F, 8, 8, 8, var1);
 		this.bipedHead.setRotationPoint(0.0F, 0.0F + var2, 0.0F);
-		this.bipedHeadwear = new ModelRenderer(32, 0);
+		this.bipedHeadwear = new ModelRenderer(this, 32, 0);
 		this.bipedHeadwear.addBox(-4.0F, -8.0F, -4.0F, 8, 8, 8, var1 + 0.5F);
 		this.bipedHeadwear.setRotationPoint(0.0F, 0.0F + var2, 0.0F);
-		this.bipedBody = new ModelRenderer(16, 16);
+		this.bipedBody = new ModelRenderer(this, 16, 16);
 		this.bipedBody.addBox(-4.0F, 0.0F, -2.0F, 8, 12, 4, var1);
 		this.bipedBody.setRotationPoint(0.0F, 0.0F + var2, 0.0F);
-		this.bipedRightArm = new ModelRenderer(40, 16);
+		this.bipedRightArm = new ModelRenderer(this, 40, 16);
 		this.bipedRightArm.addBox(-3.0F, -2.0F, -2.0F, 4, 12, 4, var1);
 		this.bipedRightArm.setRotationPoint(-5.0F, 2.0F + var2, 0.0F);
-		this.bipedLeftArm = new ModelRenderer(40, 16);
+		this.bipedLeftArm = new ModelRenderer(this, 40, 16);
 		this.bipedLeftArm.mirror = true;
 		this.bipedLeftArm.addBox(-1.0F, -2.0F, -2.0F, 4, 12, 4, var1);
 		this.bipedLeftArm.setRotationPoint(5.0F, 2.0F + var2, 0.0F);
-		this.bipedRightLeg = new ModelRenderer(0, 16);
+		this.bipedRightLeg = new ModelRenderer(this, 0, 16);
 		this.bipedRightLeg.addBox(-2.0F, 0.0F, -2.0F, 4, 12, 4, var1);
 		this.bipedRightLeg.setRotationPoint(-2.0F, 12.0F + var2, 0.0F);
-		this.bipedLeftLeg = new ModelRenderer(0, 16);
+		this.bipedLeftLeg = new ModelRenderer(this, 0, 16);
 		this.bipedLeftLeg.mirror = true;
 		this.bipedLeftLeg.addBox(-2.0F, 0.0F, -2.0F, 4, 12, 4, var1);
 		this.bipedLeftLeg.setRotationPoint(2.0F, 12.0F + var2, 0.0F);
@@ -98,8 +99,10 @@
 
 		this.bipedRightArm.rotateAngleY = 0.0F;
 		this.bipedLeftArm.rotateAngleY = 0.0F;
+		float var7;
+		float var8;
 		if(this.onGround > -9990.0F) {
-			float var7 = this.onGround;
+			var7 = this.onGround;
 			this.bipedBody.rotateAngleY = MathHelper.sin(MathHelper.sqrt_float(var7) * (float)Math.PI * 2.0F) * 0.2F;
 			this.bipedRightArm.rotationPointZ = MathHelper.sin(this.bipedBody.rotateAngleY) * 5.0F;
 			this.bipedRightArm.rotationPointX = -MathHelper.cos(this.bipedBody.rotateAngleY) * 5.0F;
@@ -112,7 +115,7 @@
 			var7 *= var7;
 			var7 *= var7;
 			var7 = 1.0F - var7;
-			float var8 = MathHelper.sin(var7 * (float)Math.PI);
+			var8 = MathHelper.sin(var7 * (float)Math.PI);
 			float var9 = MathHelper.sin(this.onGround * (float)Math.PI) * -(this.bipedHead.rotateAngleX - 0.7F) * (12.0F / 16.0F);
 			this.bipedRightArm.rotateAngleX = (float)((double)this.bipedRightArm.rotateAngleX - ((double)var8 * 1.2D + (double)var9));
 			this.bipedRightArm.rotateAngleY += this.bipedBody.rotateAngleY * 2.0F;
@@ -143,6 +146,22 @@
 		this.bipedLeftArm.rotateAngleZ -= MathHelper.cos(var3 * 0.09F) * 0.05F + 0.05F;
 		this.bipedRightArm.rotateAngleX += MathHelper.sin(var3 * 0.067F) * 0.05F;
 		this.bipedLeftArm.rotateAngleX -= MathHelper.sin(var3 * 0.067F) * 0.05F;
+		if(this.field_40333_u) {
+			var7 = 0.0F;
+			var8 = 0.0F;
+			this.bipedRightArm.rotateAngleZ = 0.0F;
+			this.bipedLeftArm.rotateAngleZ = 0.0F;
+			this.bipedRightArm.rotateAngleY = -(0.1F - var7 * 0.6F) + this.bipedHead.rotateAngleY;
+			this.bipedLeftArm.rotateAngleY = 0.1F - var7 * 0.6F + this.bipedHead.rotateAngleY + 0.4F;
+			this.bipedRightArm.rotateAngleX = (float)Math.PI * -0.5F + this.bipedHead.rotateAngleX;
+			this.bipedLeftArm.rotateAngleX = (float)Math.PI * -0.5F + this.bipedHead.rotateAngleX;
+			this.bipedRightArm.rotateAngleX -= var7 * 1.2F - var8 * 0.4F;
+			this.bipedLeftArm.rotateAngleX -= var7 * 1.2F - var8 * 0.4F;
+			this.bipedRightArm.rotateAngleZ += MathHelper.cos(var3 * 0.09F) * 0.05F + 0.05F;
+			this.bipedLeftArm.rotateAngleZ -= MathHelper.cos(var3 * 0.09F) * 0.05F + 0.05F;
+			this.bipedRightArm.rotateAngleX += MathHelper.sin(var3 * 0.067F) * 0.05F;
+			this.bipedLeftArm.rotateAngleX -= MathHelper.sin(var3 * 0.067F) * 0.05F;
+		}
 	}
 
 	public void renderEars(float var1) {
--- net/minecraft/src/PlayerController.java
+++ net/minecraft/src/PlayerController.java
@@ -97,4 +97,8 @@
 		var2.craftingInventory.onCraftGuiClosed(var2);
 		var2.craftingInventory = var2.inventorySlots;
 	}
+
+	public void onStoppedUsingItem(EntityPlayer var1) {
+		var1.stopUsingItem();
+	}
 }
--- net/minecraft/src/ItemRenderer.java
+++ net/minecraft/src/ItemRenderer.java
@@ -152,6 +152,8 @@
 		}
 
 		float var14;
+		float var13;
+		float var18;
 		if(var5 != null && var5.itemID == Item.mapItem.shiftedIndex) {
 			GL11.glPushMatrix();
 			var14 = 0.8F;
@@ -185,7 +187,7 @@
 				GL11.glRotatef((float)(-65 * var21), 0.0F, 1.0F, 0.0F);
 				Render var11 = RenderManager.instance.getEntityRenderObject(this.mc.thePlayer);
 				RenderPlayer var12 = (RenderPlayer)var11;
-				float var13 = 1.0F;
+				var13 = 1.0F;
 				GL11.glScalef(var13, var13, var13);
 				var12.drawFirstPersonHand();
 				GL11.glPopMatrix();
@@ -193,7 +195,7 @@
 
 			var9 = var3.getSwingProgress(var1);
 			var10 = MathHelper.sin(var9 * var9 * (float)Math.PI);
-			float var18 = MathHelper.sin(MathHelper.sqrt_float(var9) * (float)Math.PI);
+			var18 = MathHelper.sin(MathHelper.sqrt_float(var9) * (float)Math.PI);
 			GL11.glRotatef(-var10 * 20.0F, 0.0F, 1.0F, 0.0F);
 			GL11.glRotatef(-var18 * 20.0F, 0.0F, 0.0F, 1.0F);
 			GL11.glRotatef(-var18 * 80.0F, 1.0F, 0.0F, 0.0F);
@@ -220,10 +222,30 @@
 		} else if(var5 != null) {
 			GL11.glPushMatrix();
 			var14 = 0.8F;
-			var8 = var3.getSwingProgress(var1);
-			var9 = MathHelper.sin(var8 * (float)Math.PI);
-			var10 = MathHelper.sin(MathHelper.sqrt_float(var8) * (float)Math.PI);
-			GL11.glTranslatef(-var10 * 0.4F, MathHelper.sin(MathHelper.sqrt_float(var8) * (float)Math.PI * 2.0F) * 0.2F, -var9 * 0.2F);
+			float var12;
+			if(var3.getItemInUseCount() > 0) {
+				EnumAction var17 = var5.getItemUseAction();
+				if(var17 == EnumAction.eat) {
+					var9 = (float)var3.getItemInUseCount() - var1 + 1.0F;
+					var10 = 1.0F - var9 / (float)var5.getMaxItemUseDuration();
+					var12 = 1.0F - var10;
+					var12 = var12 * var12 * var12;
+					var12 = var12 * var12 * var12;
+					var12 = var12 * var12 * var12;
+					var13 = 1.0F - var12;
+					GL11.glTranslatef(0.0F, MathHelper.abs(MathHelper.cos(var9 / 4.0F * (float)Math.PI) * 0.1F) * (float)((double)var10 > 0.2D ? 1 : 0), 0.0F);
+					GL11.glTranslatef(var13 * 0.6F, -var13 * 0.5F, 0.0F);
+					GL11.glRotatef(var13 * 90.0F, 0.0F, 1.0F, 0.0F);
+					GL11.glRotatef(var13 * 10.0F, 1.0F, 0.0F, 0.0F);
+					GL11.glRotatef(var13 * 30.0F, 0.0F, 0.0F, 1.0F);
+				}
+			} else {
+				var8 = var3.getSwingProgress(var1);
+				var9 = MathHelper.sin(var8 * (float)Math.PI);
+				var10 = MathHelper.sin(MathHelper.sqrt_float(var8) * (float)Math.PI);
+				GL11.glTranslatef(-var10 * 0.4F, MathHelper.sin(MathHelper.sqrt_float(var8) * (float)Math.PI * 2.0F) * 0.2F, -var9 * 0.2F);
+			}
+
 			GL11.glTranslatef(0.7F * var14, -0.65F * var14 - (1.0F - var2) * 0.6F, -0.9F * var14);
 			GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
 			GL11.glEnable(GL12.GL_RESCALE_NORMAL);
@@ -235,6 +257,37 @@
 			GL11.glRotatef(-var10 * 80.0F, 1.0F, 0.0F, 0.0F);
 			var8 = 0.4F;
 			GL11.glScalef(var8, var8, var8);
+
+			if(var3.getItemInUseCount() > 0) {
+				EnumAction var17 = var5.getItemUseAction();
+				if(var17 == EnumAction.bow) {
+					GL11.glRotatef(-18.0F, 0.0F, 0.0F, 1.0F);
+					GL11.glRotatef(-12.0F, 0.0F, 1.0F, 0.0F);
+					GL11.glRotatef(-8.0F, 1.0F, 0.0F, 0.0F);
+					GL11.glTranslatef(-0.9F, 0.2F, 0.0F);
+					var10 = (float)var5.getMaxItemUseDuration() - ((float)var3.getItemInUseCount() - var1 + 1.0F);
+					var18 = var10 / 20.0F;
+					var18 = (var18 * var18 + var18 * 2.0F) / 3.0F;
+					if(var18 > 1.0F) {
+						var18 = 1.0F;
+					}
+
+					if(var18 > 0.1F) {
+						GL11.glTranslatef(0.0F, MathHelper.sin((var10 - 0.1F) * 1.3F) * 0.01F * (var18 - 0.1F), 0.0F);
+					}
+
+					GL11.glTranslatef(0.0F, 0.0F, var18 * 0.1F);
+					GL11.glRotatef(-335.0F, 0.0F, 0.0F, 1.0F);
+					GL11.glRotatef(-50.0F, 0.0F, 1.0F, 0.0F);
+					GL11.glTranslatef(0.0F, 0.5F, 0.0F);
+					var12 = 1.0F + var18 * 0.2F;
+					GL11.glScalef(1.0F, 1.0F, var12);
+					GL11.glTranslatef(0.0F, -0.5F, 0.0F);
+					GL11.glRotatef(50.0F, 0.0F, 1.0F, 0.0F);
+					GL11.glRotatef(335.0F, 0.0F, 0.0F, 1.0F);
+				}
+			}
+
 			if(var5.getItem().shouldRotateAroundWhenRendering()) {
 				GL11.glRotatef(180.0F, 0.0F, 1.0F, 0.0F);
 			}
--- net/minecraft/src/ModelWolf.java
+++ net/minecraft/src/ModelWolf.java
@@ -18,37 +18,37 @@
 	public ModelWolf() {
 		float var1 = 0.0F;
 		float var2 = 13.5F;
-		this.wolfHeadMain = new ModelRenderer(0, 0);
+		this.wolfHeadMain = new ModelRenderer(this, 0, 0);
 		this.wolfHeadMain.addBox(-3.0F, -3.0F, -2.0F, 6, 6, 4, var1);
 		this.wolfHeadMain.setRotationPoint(-1.0F, var2, -7.0F);
-		this.wolfBody = new ModelRenderer(18, 14);
+		this.wolfBody = new ModelRenderer(this, 18, 14);
 		this.wolfBody.addBox(-4.0F, -2.0F, -3.0F, 6, 9, 6, var1);
 		this.wolfBody.setRotationPoint(0.0F, 14.0F, 2.0F);
-		this.wolfMane = new ModelRenderer(21, 0);
+		this.wolfMane = new ModelRenderer(this, 21, 0);
 		this.wolfMane.addBox(-4.0F, -3.0F, -3.0F, 8, 6, 7, var1);
 		this.wolfMane.setRotationPoint(-1.0F, 14.0F, 2.0F);
-		this.wolfLeg1 = new ModelRenderer(0, 18);
+		this.wolfLeg1 = new ModelRenderer(this, 0, 18);
 		this.wolfLeg1.addBox(-1.0F, 0.0F, -1.0F, 2, 8, 2, var1);
 		this.wolfLeg1.setRotationPoint(-2.5F, 16.0F, 7.0F);
-		this.wolfLeg2 = new ModelRenderer(0, 18);
+		this.wolfLeg2 = new ModelRenderer(this, 0, 18);
 		this.wolfLeg2.addBox(-1.0F, 0.0F, -1.0F, 2, 8, 2, var1);
 		this.wolfLeg2.setRotationPoint(0.5F, 16.0F, 7.0F);
-		this.wolfLeg3 = new ModelRenderer(0, 18);
+		this.wolfLeg3 = new ModelRenderer(this, 0, 18);
 		this.wolfLeg3.addBox(-1.0F, 0.0F, -1.0F, 2, 8, 2, var1);
 		this.wolfLeg3.setRotationPoint(-2.5F, 16.0F, -4.0F);
-		this.wolfLeg4 = new ModelRenderer(0, 18);
+		this.wolfLeg4 = new ModelRenderer(this, 0, 18);
 		this.wolfLeg4.addBox(-1.0F, 0.0F, -1.0F, 2, 8, 2, var1);
 		this.wolfLeg4.setRotationPoint(0.5F, 16.0F, -4.0F);
-		this.wolfTail = new ModelRenderer(9, 18);
+		this.wolfTail = new ModelRenderer(this, 9, 18);
 		this.wolfTail.addBox(-1.0F, 0.0F, -1.0F, 2, 8, 2, var1);
 		this.wolfTail.setRotationPoint(-1.0F, 12.0F, 8.0F);
-		this.wolfRightEar = new ModelRenderer(16, 14);
+		this.wolfRightEar = new ModelRenderer(this, 16, 14);
 		this.wolfRightEar.addBox(-3.0F, -5.0F, 0.0F, 2, 2, 1, var1);
 		this.wolfRightEar.setRotationPoint(-1.0F, var2, -7.0F);
-		this.wolfLeftEar = new ModelRenderer(16, 14);
+		this.wolfLeftEar = new ModelRenderer(this, 16, 14);
 		this.wolfLeftEar.addBox(1.0F, -5.0F, 0.0F, 2, 2, 1, var1);
 		this.wolfLeftEar.setRotationPoint(-1.0F, var2, -7.0F);
-		this.wolfSnout = new ModelRenderer(0, 10);
+		this.wolfSnout = new ModelRenderer(this, 0, 10);
 		this.wolfSnout.addBox(-2.0F, 0.0F, -5.0F, 3, 3, 4, var1);
 		this.wolfSnout.setRotationPoint(-0.5F, var2, -7.0F);
 	}
--- net/minecraft/src/ModelGhast.java
+++ net/minecraft/src/ModelGhast.java
@@ -8,13 +8,13 @@
 
 	public ModelGhast() {
 		byte var1 = -16;
-		this.body = new ModelRenderer(0, 0);
+		this.body = new ModelRenderer(this, 0, 0);
 		this.body.addBox(-8.0F, -8.0F, -8.0F, 16, 16, 16);
 		this.body.rotationPointY += (float)(24 + var1);
 		Random var2 = new Random(1660L);
 
 		for(int var3 = 0; var3 < this.tentacles.length; ++var3) {
-			this.tentacles[var3] = new ModelRenderer(0, 0);
+			this.tentacles[var3] = new ModelRenderer(this, 0, 0);
 			float var4 = (((float)(var3 % 3) - (float)(var3 / 3 % 2) * 0.5F + 0.25F) / 2.0F * 2.0F - 1.0F) * 5.0F;
 			float var5 = ((float)(var3 / 3) / 2.0F * 2.0F - 1.0F) * 5.0F;
 			int var6 = var2.nextInt(7) + 8;
--- net/minecraft/src/RenderBiped.java
+++ net/minecraft/src/RenderBiped.java
@@ -24,6 +24,13 @@
 				GL11.glRotatef(20.0F, 1.0F, 0.0F, 0.0F);
 				GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
 				GL11.glScalef(var4, -var4, var4);
+			} else if(var3.itemID == Item.bow.shiftedIndex) {
+				var4 = 10.0F / 16.0F;
+				GL11.glTranslatef(0.0F, 2.0F / 16.0F, 5.0F / 16.0F);
+				GL11.glRotatef(-20.0F, 0.0F, 1.0F, 0.0F);
+				GL11.glScalef(var4, -var4, var4);
+				GL11.glRotatef(-100.0F, 1.0F, 0.0F, 0.0F);
+				GL11.glRotatef(45.0F, 0.0F, 1.0F, 0.0F);
 			} else if(Item.itemsList[var3.itemID].isFull3D()) {
 				var4 = 10.0F / 16.0F;
 				GL11.glTranslatef(0.0F, 3.0F / 16.0F, 0.0F);
--- net/minecraft/src/EntityLiving.java
+++ net/minecraft/src/EntityLiving.java
@@ -39,6 +39,8 @@
 	public float field_705_Q;
 	public float field_704_R;
 	public float field_703_S;
+	public int field_35172_bP = 0;
+	public int field_35173_bQ = 0;
 	protected int newPosRotationIncrements;
 	protected double newPosX;
 	protected double newPosY;
@@ -202,6 +204,16 @@
 
 	public void onUpdate() {
 		super.onUpdate();
+		if(this.field_35172_bP > 0) {
+			if(this.field_35173_bQ <= 0) {
+				this.field_35173_bQ = 600;
+			}
+
+			--this.field_35173_bQ;
+			if(this.field_35173_bQ <= 0) {
+				--this.field_35172_bP;
+			}
+		}
 		this.onLivingUpdate();
 		double var1 = this.posX - this.prevPosX;
 		double var3 = this.posZ - this.prevPosZ;
--- /dev/null
+++ net/minecraft/src/EntityBreakingFX.java
@@ -1,0 +1,42 @@
+package net.minecraft.src;
+
+public class EntityBreakingFX extends EntityFX {
+	public EntityBreakingFX(World var1, double var2, double var4, double var6, Item var8) {
+		super(var1, var2, var4, var6, 0.0D, 0.0D, 0.0D);
+		this.particleTextureIndex = var8.getIconFromDamage(0);
+		this.particleRed = this.particleGreen = this.particleBlue = 1.0F;
+		this.particleGravity = Block.blockSnow.blockParticleGravity;
+		this.particleScale /= 2.0F;
+	}
+
+	public EntityBreakingFX(World var1, double var2, double var4, double var6, double var8, double var10, double var12, Item var14) {
+		this(var1, var2, var4, var6, var14);
+		this.motionX *= (double)0.1F;
+		this.motionY *= (double)0.1F;
+		this.motionZ *= (double)0.1F;
+		this.motionX += var8;
+		this.motionY += var10;
+		this.motionZ += var12;
+	}
+
+	public int getFXLayer() {
+		return 2;
+	}
+
+	public void renderParticle(Tessellator var1, float var2, float var3, float var4, float var5, float var6, float var7) {
+		float var8 = ((float)(this.particleTextureIndex % 16) + this.particleTextureJitterX / 4.0F) / 16.0F;
+		float var9 = var8 + 0.999F / 64.0F;
+		float var10 = ((float)(this.particleTextureIndex / 16) + this.particleTextureJitterY / 4.0F) / 16.0F;
+		float var11 = var10 + 0.999F / 64.0F;
+		float var12 = 0.1F * this.particleScale;
+		float var13 = (float)(this.prevPosX + (this.posX - this.prevPosX) * (double)var2 - interpPosX);
+		float var14 = (float)(this.prevPosY + (this.posY - this.prevPosY) * (double)var2 - interpPosY);
+		float var15 = (float)(this.prevPosZ + (this.posZ - this.prevPosZ) * (double)var2 - interpPosZ);
+		float var16 = 1.0F;
+		var1.setColorOpaque_F(var16 * this.particleRed, var16 * this.particleGreen, var16 * this.particleBlue);
+		var1.addVertexWithUV((double)(var13 - var3 * var12 - var6 * var12), (double)(var14 - var4 * var12), (double)(var15 - var5 * var12 - var7 * var12), (double)var8, (double)var11);
+		var1.addVertexWithUV((double)(var13 - var3 * var12 + var6 * var12), (double)(var14 + var4 * var12), (double)(var15 - var5 * var12 + var7 * var12), (double)var8, (double)var10);
+		var1.addVertexWithUV((double)(var13 + var3 * var12 + var6 * var12), (double)(var14 + var4 * var12), (double)(var15 + var5 * var12 + var7 * var12), (double)var9, (double)var10);
+		var1.addVertexWithUV((double)(var13 + var3 * var12 - var6 * var12), (double)(var14 - var4 * var12), (double)(var15 + var5 * var12 - var7 * var12), (double)var9, (double)var11);
+	}
+}
--- net/minecraft/src/RenderItem.java
+++ net/minecraft/src/RenderItem.java
@@ -163,9 +163,13 @@
 		GL11.glEnable(GL11.GL_CULL_FACE);
 	}
 
-	public void renderItemIntoGUI(FontRenderer var1, RenderEngine var2, ItemStack var3, int var4, int var5) {
+	public void renderItemIntoGUI(FontRenderer var1, RenderEngine var2, ItemStack var3, int var4, int var5, EntityPlayer var6) {
 		if(var3 != null) {
-			this.drawItemIntoGui(var1, var2, var3.itemID, var3.getItemDamage(), var3.getIconIndex(), var4, var5);
+			if(var6 != null) {
+				this.drawItemIntoGui(var1, var2, var3.itemID, var3.getItemDamage(), var6.getItemIcon(var3), var4, var5);
+			} else {
+				this.drawItemIntoGui(var1, var2, var3.itemID, var3.getItemDamage(), var3.getIconIndex(), var4, var5);
+			}
 		}
 	}
 
--- net/minecraft/src/ModelSheep1.java
+++ net/minecraft/src/ModelSheep1.java
@@ -3,23 +3,23 @@
 public class ModelSheep1 extends ModelQuadruped {
 	public ModelSheep1() {
 		super(12, 0.0F);
-		this.head = new ModelRenderer(0, 0);
+		this.head = new ModelRenderer(this, 0, 0);
 		this.head.addBox(-3.0F, -4.0F, -4.0F, 6, 6, 6, 0.6F);
 		this.head.setRotationPoint(0.0F, 6.0F, -8.0F);
-		this.body = new ModelRenderer(28, 8);
+		this.body = new ModelRenderer(this, 28, 8);
 		this.body.addBox(-4.0F, -10.0F, -7.0F, 8, 16, 6, 1.75F);
 		this.body.setRotationPoint(0.0F, 5.0F, 2.0F);
 		float var1 = 0.5F;
-		this.leg1 = new ModelRenderer(0, 16);
+		this.leg1 = new ModelRenderer(this, 0, 16);
 		this.leg1.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg1.setRotationPoint(-3.0F, 12.0F, 7.0F);
-		this.leg2 = new ModelRenderer(0, 16);
+		this.leg2 = new ModelRenderer(this, 0, 16);
 		this.leg2.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg2.setRotationPoint(3.0F, 12.0F, 7.0F);
-		this.leg3 = new ModelRenderer(0, 16);
+		this.leg3 = new ModelRenderer(this, 0, 16);
 		this.leg3.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg3.setRotationPoint(-3.0F, 12.0F, -5.0F);
-		this.leg4 = new ModelRenderer(0, 16);
+		this.leg4 = new ModelRenderer(this, 0, 16);
 		this.leg4.addBox(-2.0F, 0.0F, -2.0F, 4, 6, 4, var1);
 		this.leg4.setRotationPoint(3.0F, 12.0F, -5.0F);
 	}
--- net/minecraft/src/GameSettings.java
+++ net/minecraft/src/GameSettings.java
@@ -48,6 +48,7 @@
 	public boolean field_22273_E = false;
 	public float field_22272_F = 1.0F;
 	public float field_22271_G = 1.0F;
+	public float fovSetting = 0.0F;
 	public int guiScale = 0;
 
 	public GameSettings(Minecraft var1, File var2) {
@@ -84,6 +85,10 @@
 			this.mc.sndManager.onSoundOptionsChanged();
 		}
 
+		if(var1 == EnumOptions.FOV) {
+			this.fovSetting = var2;
+		}
+
 		if(var1 == EnumOptions.SENSITIVITY) {
 			this.mouseSensitivity = var2;
 		}
@@ -139,7 +144,7 @@
 	}
 
 	public float getOptionFloatValue(EnumOptions var1) {
-		return var1 == EnumOptions.MUSIC ? this.musicVolume : (var1 == EnumOptions.SOUND ? this.soundVolume : (var1 == EnumOptions.SENSITIVITY ? this.mouseSensitivity : 0.0F));
+		return var1 == EnumOptions.FOV ? this.fovSetting : (var1 == EnumOptions.MUSIC ? this.musicVolume : (var1 == EnumOptions.SOUND ? this.soundVolume : (var1 == EnumOptions.SENSITIVITY ? this.mouseSensitivity : 0.0F)));
 	}
 
 	public boolean getOptionOrdinalValue(EnumOptions var1) {
@@ -164,7 +169,7 @@
 		String var3 = var2.translateKey(var1.getEnumString()) + ": ";
 		if(var1.getEnumFloat()) {
 			float var5 = this.getOptionFloatValue(var1);
-			return var1 == EnumOptions.SENSITIVITY ? (var5 == 0.0F ? var3 + var2.translateKey("options.sensitivity.min") : (var5 == 1.0F ? var3 + var2.translateKey("options.sensitivity.max") : var3 + (int)(var5 * 200.0F) + "%")) : (var5 == 0.0F ? var3 + var2.translateKey("options.off") : var3 + (int)(var5 * 100.0F) + "%");
+			return var1 == EnumOptions.SENSITIVITY ? (var5 == 0.0F ? var3 + var2.translateKey("options.sensitivity.min") : (var5 == 1.0F ? var3 + var2.translateKey("options.sensitivity.max") : var3 + (int)(var5 * 200.0F) + "%")) : (var1 == EnumOptions.FOV ? (var5 == 0.0F ? var3 + var2.translateKey("options.fov.min") : (var5 == 1.0F ? var3 + var2.translateKey("options.fov.max") : var3 + (int)(70.0F + var5 * 40.0F))) : (var5 == 0.0F ? var3 + var2.translateKey("options.off") : var3 + (int)(var5 * 100.0F) + "%"));
 		} else if(var1.getEnumBoolean()) {
 			boolean var4 = this.getOptionOrdinalValue(var1);
 			return var4 ? var3 + var2.translateKey("options.on") : var3 + var2.translateKey("options.off");
@@ -202,6 +207,10 @@
 					if(var3[0].equals("mouseSensitivity")) {
 						this.mouseSensitivity = this.parseFloat(var3[1]);
 					}
+					
+					if(var3[0].equals("fov")) {
+						this.fovSetting = this.parseFloat(var3[1]);
+					}
 
 					if(var3[0].equals("invertYMouse")) {
 						this.invertMouse = var3[1].equals("true");
@@ -278,6 +287,7 @@
 			var1.println("sound:" + this.soundVolume);
 			var1.println("invertYMouse:" + this.invertMouse);
 			var1.println("mouseSensitivity:" + this.mouseSensitivity);
+			var1.println("fov:" + this.fovSetting);
 			var1.println("viewDistance:" + this.renderDistance);
 			var1.println("guiScale:" + this.guiScale);
 			var1.println("bobView:" + this.viewBobbing);
--- net/minecraft/src/ModelBase.java
+++ net/minecraft/src/ModelBase.java
@@ -1,8 +1,13 @@
 package net.minecraft.src;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
 public abstract class ModelBase {
 	public float onGround;
 	public boolean isRiding = false;
+	public List field_35394_j = new ArrayList();
 
 	public void render(float var1, float var2, float var3, float var4, float var5, float var6) {
 	}
@@ -11,5 +16,9 @@
 	}
 
 	public void setLivingAnimations(EntityLiving var1, float var2, float var3, float var4) {
+	}
+
+	public ModelRenderer func_35393_a(Random var1) {
+		return (ModelRenderer)this.field_35394_j.get(var1.nextInt(this.field_35394_j.size()));
 	}
 }
