--- net/minecraft/src/RenderBlocks.java
+++ net/minecraft/src/RenderBlocks.java
@@ -1470,151 +1470,750 @@
 	}
 
 	public boolean renderBlockFluids(Block var1, int var2, int var3, int var4) {
-		Tessellator var5 = Tessellator.instance;
 		int var6 = var1.colorMultiplier(this.blockAccess, var2, var3, var4);
 		float var7 = (float)(var6 >> 16 & 255) / 255.0F;
 		float var8 = (float)(var6 >> 8 & 255) / 255.0F;
 		float var9 = (float)(var6 & 255) / 255.0F;
-		boolean var10 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 + 1, var4, 1);
-		boolean var11 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 - 1, var4, 0);
-		boolean[] var12 = new boolean[]{var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 - 1, 2), var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 + 1, 3), var1.shouldSideBeRendered(this.blockAccess, var2 - 1, var3, var4, 4), var1.shouldSideBeRendered(this.blockAccess, var2 + 1, var3, var4, 5)};
-		if(!var10 && !var11 && !var12[0] && !var12[1] && !var12[2] && !var12[3]) {
+		if(EntityRenderer.field_28135_a) {
+			float var10 = (var7 * 30.0F + var8 * 59.0F + var9 * 11.0F) / 100.0F;
+			float var11 = (var7 * 30.0F + var8 * 70.0F) / 100.0F;
+			float var12 = (var7 * 30.0F + var9 * 70.0F) / 100.0F;
+			var7 = var10;
+			var8 = var11;
+			var9 = var12;
+		}
+
+		return Minecraft.isAmbientOcclusionEnabled() ? this.renderBlockFluidsWithAmbientOcclusion(var1, var2, var3, var4, var7, var8, var9) : this.renderBlockFluidsWithColorMultiplier(var1, var2, var3, var4, var7, var8, var9);
+	}
+
+	public boolean renderBlockFluidsWithAmbientOcclusion(Block var1, int var2, int var3, int var4, float var5, float var6, float var7) {
+		Tessellator var8 = Tessellator.instance;
+		float var9 = 0.0F;
+		float var10 = 0.0F;
+		float var11 = 0.0F;
+		float var12 = 0.0F;
+		boolean var13 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 + 1, var4, 1);
+		boolean var14 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 - 1, var4, 0);
+		boolean var15 = var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 - 1, 2);
+		boolean var16 = var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 + 1, 3);
+		boolean var17 = var1.shouldSideBeRendered(this.blockAccess, var2 - 1, var3, var4, 4);
+		boolean var18 = var1.shouldSideBeRendered(this.blockAccess, var2 + 1, var3, var4, 5);
+		if(!this.renderAllFaces && !var13 && !var14 && !var15 && !var16 && !var17 && !var18) {
 			return false;
 		} else {
-			boolean var13 = false;
-			float var14 = 0.5F;
-			float var15 = 1.0F;
-			float var16 = 0.8F;
-			float var17 = 0.6F;
-			double var18 = 0.0D;
-			double var20 = 1.0D;
-			Material var22 = var1.blockMaterial;
-			int var23 = this.blockAccess.getBlockMetadata(var2, var3, var4);
-			float var24 = this.func_1224_a(var2, var3, var4, var22);
-			float var25 = this.func_1224_a(var2, var3, var4 + 1, var22);
-			float var26 = this.func_1224_a(var2 + 1, var3, var4 + 1, var22);
-			float var27 = this.func_1224_a(var2 + 1, var3, var4, var22);
-			int var28;
-			int var31;
+			boolean var19 = false;
+			Material var20 = var1.blockMaterial;
+			int var21 = this.blockAccess.getBlockMetadata(var2, var3, var4);
+			float var22 = this.func_1224_a(var2, var3, var4, var20);
+			float var23 = this.func_1224_a(var2, var3, var4 + 1, var20);
+			float var24 = this.func_1224_a(var2 + 1, var3, var4 + 1, var20);
+			float var25 = this.func_1224_a(var2 + 1, var3, var4, var20);
+			boolean var26;
+			boolean var27;
+			boolean var28;
+			boolean var29;
+			float var30;
+			float var34;
+			float var35;
+			int var47;
+			int var50;
+			int var51;
+			if(this.renderAllFaces || var13) {
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var2, var3, var4);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					this.field_22368_v = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4);
+					this.field_22364_z = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4);
+					this.field_22366_x = var1.getBlockBrightness(this.blockAccess, var2, var3, var4 - 1);
+					this.field_22362_A = var1.getBlockBrightness(this.blockAccess, var2, var3, var4 + 1);
+					var26 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 + 1, var3, var4)];
+					var27 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 - 1, var3, var4)];
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3, var4 + 1)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3, var4 - 1)];
+					if(!var29 && !var27) {
+						this.field_22369_u = this.field_22368_v;
+					} else {
+						this.field_22369_u = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4 - 1);
+					}
+
+					if(!var29 && !var26) {
+						this.field_22365_y = this.field_22364_z;
+					} else {
+						this.field_22365_y = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4 - 1);
+					}
+
+					if(!var28 && !var27) {
+						this.field_22367_w = this.field_22368_v;
+					} else {
+						this.field_22367_w = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4 + 1);
+					}
+
+					if(!var28 && !var26) {
+						this.field_22360_B = this.field_22364_z;
+					} else {
+						this.field_22360_B = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4 + 1);
+					}
+
+					var30 = var1.getBlockBrightness(this.blockAccess, var2, var3, var4);
+					var12 = (this.field_22367_w + this.field_22368_v + this.field_22362_A + var30) / 4.0F;
+					var9 = (this.field_22362_A + var30 + this.field_22360_B + this.field_22364_z) / 4.0F;
+					var10 = (var30 + this.field_22366_x + this.field_22364_z + this.field_22365_y) / 4.0F;
+					var11 = (this.field_22368_v + this.field_22369_u + var30 + this.field_22366_x) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				var19 = true;
+				var47 = var1.getBlockTextureFromSideAndMetadata(1, var21);
+				float var48 = (float)BlockFluid.func_293_a(this.blockAccess, var2, var3, var4, var20);
+				if(var48 > -999.0F) {
+					var47 = var1.getBlockTextureFromSideAndMetadata(2, var21);
+				}
+
+				var50 = (var47 & 15) << 4;
+				var51 = var47 & 240;
+				double var32;
+				double var52;
+				if(var48 < -999.0F) {
+					var52 = ((double)var50 + 8.0D) / 256.0D;
+					var32 = ((double)var51 + 8.0D) / 256.0D;
+					var34 = 0.0F;
+					var35 = 0.03125F;
+				} else {
+					var52 = (double)((float)(var50 + 16) / 256.0F);
+					var32 = (double)((float)(var51 + 16) / 256.0F);
+					var34 = MathHelper.sin(var48) * 8.0F / 256.0F;
+					var35 = MathHelper.cos(var48) * 8.0F / 256.0F;
+				}
+
+				var8.setColorOpaque_F(this.colorRedBottomRight, this.colorGreenBottomRight, this.colorBlueBottomRight);
+				var8.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var22), (double)(var4 + 0), var52 - (double)var35 - (double)var34, var32 - (double)var35 + (double)var34);
+				var8.setColorOpaque_F(this.colorRedTopRight, this.colorGreenTopRight, this.colorBlueTopRight);
+				var8.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var23), (double)(var4 + 1), var52 - (double)var35 + (double)var34, var32 + (double)var35 + (double)var34);
+				var8.setColorOpaque_F(this.colorRedTopLeft, this.colorGreenTopLeft, this.colorBlueTopLeft);
+				var8.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var24), (double)(var4 + 1), var52 + (double)var35 + (double)var34, var32 + (double)var35 - (double)var34);
+				var8.setColorOpaque_F(this.colorRedBottomLeft, this.colorGreenBottomLeft, this.colorBlueBottomLeft);
+				var8.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var25), (double)(var4 + 0), var52 + (double)var35 - (double)var34, var32 - (double)var35 - (double)var34);
+			}
+
+			if(this.renderAllFaces || var14) {
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var2, var3 - 1, var4);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					--var3;
+					this.field_22376_n = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4);
+					this.field_22374_p = var1.getBlockBrightness(this.blockAccess, var2, var3, var4 - 1);
+					this.field_22373_q = var1.getBlockBrightness(this.blockAccess, var2, var3, var4 + 1);
+					this.field_22371_s = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4);
+					var26 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 + 1, var3, var4)];
+					var27 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 - 1, var3, var4)];
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3, var4 + 1)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3, var4 - 1)];
+					if(!var29 && !var27) {
+						this.field_22377_m = this.field_22376_n;
+					} else {
+						this.field_22377_m = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4 - 1);
+					}
+
+					if(!var28 && !var27) {
+						this.field_22375_o = this.field_22376_n;
+					} else {
+						this.field_22375_o = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var4 + 1);
+					}
+
+					if(!var29 && !var26) {
+						this.field_22372_r = this.field_22371_s;
+					} else {
+						this.field_22372_r = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4 - 1);
+					}
+
+					if(!var28 && !var26) {
+						this.field_22370_t = this.field_22371_s;
+					} else {
+						this.field_22370_t = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var4 + 1);
+					}
+
+					var30 = var1.getBlockBrightness(this.blockAccess, var2, var3, var4);
+					++var3;
+					var9 = (this.field_22375_o + this.field_22376_n + this.field_22373_q + var30) / 4.0F;
+					var12 = (this.field_22373_q + var30 + this.field_22370_t + this.field_22371_s) / 4.0F;
+					var11 = (var30 + this.field_22374_p + this.field_22371_s + this.field_22372_r) / 4.0F;
+					var10 = (this.field_22376_n + this.field_22377_m + var30 + this.field_22374_p) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5 * 0.5F;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6 * 0.5F;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7 * 0.5F;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				this.renderBottomFace(var1, (double)var2, (double)var3, (double)var4, var1.getBlockTextureFromSide(0));
+				var19 = true;
+			}
+
+			boolean var31;
+			float var33;
 			float var36;
-			float var37;
+			double var37;
+			double var39;
+			double var41;
+			double var43;
+			double var45;
+			int var49;
+			boolean var53;
+			int var54;
+			float var55;
+			if(this.renderAllFaces || var15) {
+				var49 = var4 - 1;
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var2, var3, var49);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					this.field_22358_C = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var49);
+					this.field_22374_p = var1.getBlockBrightness(this.blockAccess, var2, var3 - 1, var49);
+					this.field_22366_x = var1.getBlockBrightness(this.blockAccess, var2, var3 + 1, var49);
+					this.field_22356_D = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var49);
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 + 1, var3, var49)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 - 1, var3, var49)];
+					var53 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3 + 1, var49)];
+					var31 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3 - 1, var49)];
+					if(!var29 && !var31) {
+						this.field_22377_m = this.field_22358_C;
+					} else {
+						this.field_22377_m = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3 - 1, var49);
+					}
+
+					if(!var29 && !var53) {
+						this.field_22369_u = this.field_22358_C;
+					} else {
+						this.field_22369_u = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3 + 1, var49);
+					}
+
+					if(!var28 && !var31) {
+						this.field_22372_r = this.field_22356_D;
+					} else {
+						this.field_22372_r = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3 - 1, var49);
+					}
+
+					if(!var28 && !var53) {
+						this.field_22365_y = this.field_22356_D;
+					} else {
+						this.field_22365_y = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3 + 1, var49);
+					}
+
+					var55 = var1.getBlockBrightness(this.blockAccess, var2, var3, var49);
+					var9 = (this.field_22358_C + this.field_22369_u + var55 + this.field_22366_x) / 4.0F;
+					var10 = (var55 + this.field_22366_x + this.field_22356_D + this.field_22365_y) / 4.0F;
+					var11 = (this.field_22374_p + var55 + this.field_22372_r + this.field_22356_D) / 4.0F;
+					var12 = (this.field_22377_m + this.field_22358_C + this.field_22374_p + var55) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5 * 0.8F;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6 * 0.8F;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7 * 0.8F;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				var50 = var1.getBlockTextureFromSideAndMetadata(2, var21);
+				var51 = (var50 & 15) << 4;
+				var54 = var50 & 240;
+				var33 = (float)var2;
+				var34 = (float)(var2 + 1);
+				var35 = (float)var4;
+				var36 = (float)var4;
+				var19 = true;
+				var37 = (double)((float)(var51 + 0) / 256.0F);
+				var39 = ((double)(var51 + 16) - 0.01D) / 256.0D;
+				var41 = (double)(((float)var54 + (1.0F - var22) * 16.0F) / 256.0F);
+				var43 = (double)(((float)var54 + (1.0F - var25) * 16.0F) / 256.0F);
+				var45 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var8.setColorOpaque_F(this.colorRedTopLeft, this.colorGreenTopLeft, this.colorBlueTopLeft);
+				var8.addVertexWithUV((double)var33, (double)((float)var3 + var22), (double)var35, var37, var41);
+				var8.setColorOpaque_F(this.colorRedBottomLeft, this.colorGreenBottomLeft, this.colorBlueBottomLeft);
+				var8.addVertexWithUV((double)var34, (double)((float)var3 + var25), (double)var36, var39, var43);
+				var8.setColorOpaque_F(this.colorRedBottomRight, this.colorGreenBottomRight, this.colorBlueBottomRight);
+				var8.addVertexWithUV((double)var34, (double)(var3 + 0), (double)var36, var39, var45);
+				var8.setColorOpaque_F(this.colorRedTopRight, this.colorGreenTopRight, this.colorBlueTopRight);
+				var8.addVertexWithUV((double)var33, (double)(var3 + 0), (double)var35, var37, var45);
+			}
+
+			if(this.renderAllFaces || var16) {
+				var49 = var4 + 1;
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var2, var3, var49);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					this.field_22354_E = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3, var49);
+					this.field_22353_F = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3, var49);
+					this.field_22373_q = var1.getBlockBrightness(this.blockAccess, var2, var3 - 1, var49);
+					this.field_22362_A = var1.getBlockBrightness(this.blockAccess, var2, var3 + 1, var49);
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 + 1, var3, var49)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var2 - 1, var3, var49)];
+					var53 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3 + 1, var49)];
+					var31 = Block.canBlockGrass[this.blockAccess.getBlockId(var2, var3 - 1, var49)];
+					if(!var29 && !var31) {
+						this.field_22375_o = this.field_22354_E;
+					} else {
+						this.field_22375_o = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3 - 1, var49);
+					}
+
+					if(!var29 && !var53) {
+						this.field_22367_w = this.field_22354_E;
+					} else {
+						this.field_22367_w = var1.getBlockBrightness(this.blockAccess, var2 - 1, var3 + 1, var49);
+					}
+
+					if(!var28 && !var31) {
+						this.field_22370_t = this.field_22353_F;
+					} else {
+						this.field_22370_t = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3 - 1, var49);
+					}
+
+					if(!var28 && !var53) {
+						this.field_22360_B = this.field_22353_F;
+					} else {
+						this.field_22360_B = var1.getBlockBrightness(this.blockAccess, var2 + 1, var3 + 1, var49);
+					}
+
+					var55 = var1.getBlockBrightness(this.blockAccess, var2, var3, var49);
+					var9 = (this.field_22354_E + this.field_22367_w + var55 + this.field_22362_A) / 4.0F;
+					var12 = (var55 + this.field_22362_A + this.field_22353_F + this.field_22360_B) / 4.0F;
+					var11 = (this.field_22373_q + var55 + this.field_22370_t + this.field_22353_F) / 4.0F;
+					var10 = (this.field_22375_o + this.field_22354_E + this.field_22373_q + var55) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5 * 0.8F;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6 * 0.8F;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7 * 0.8F;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				var50 = var1.getBlockTextureFromSideAndMetadata(3, var21);
+				var51 = (var50 & 15) << 4;
+				var54 = var50 & 240;
+				var33 = (float)(var2 + 1);
+				var34 = (float)var2;
+				var35 = (float)(var4 + 1);
+				var36 = (float)(var4 + 1);
+				var19 = true;
+				var37 = (double)((float)(var51 + 0) / 256.0F);
+				var39 = ((double)(var51 + 16) - 0.01D) / 256.0D;
+				var41 = (double)(((float)var54 + (1.0F - var24) * 16.0F) / 256.0F);
+				var43 = (double)(((float)var54 + (1.0F - var23) * 16.0F) / 256.0F);
+				var45 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var8.setColorOpaque_F(this.colorRedTopRight, this.colorGreenTopRight, this.colorBlueTopRight);
+				var8.addVertexWithUV((double)var33, (double)((float)var3 + var24), (double)var35, var37, var41);
+				var8.setColorOpaque_F(this.colorRedTopLeft, this.colorGreenTopLeft, this.colorBlueTopLeft);
+				var8.addVertexWithUV((double)var34, (double)((float)var3 + var23), (double)var36, var39, var43);
+				var8.setColorOpaque_F(this.colorRedBottomLeft, this.colorGreenBottomLeft, this.colorBlueBottomLeft);
+				var8.addVertexWithUV((double)var34, (double)(var3 + 0), (double)var36, var39, var45);
+				var8.setColorOpaque_F(this.colorRedBottomRight, this.colorGreenBottomRight, this.colorBlueBottomRight);
+				var8.addVertexWithUV((double)var33, (double)(var3 + 0), (double)var35, var37, var45);
+			}
+
+			if(this.renderAllFaces || var17) {
+				var47 = var2 - 1;
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var47, var3, var4);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					this.field_22376_n = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4);
+					this.field_22358_C = var1.getBlockBrightness(this.blockAccess, var47, var3, var4 - 1);
+					this.field_22354_E = var1.getBlockBrightness(this.blockAccess, var47, var3, var4 + 1);
+					this.field_22368_v = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4);
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3 + 1, var4)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3 - 1, var4)];
+					var53 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3, var4 - 1)];
+					var31 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3, var4 + 1)];
+					if(!var53 && !var29) {
+						this.field_22377_m = this.field_22358_C;
+					} else {
+						this.field_22377_m = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4 - 1);
+					}
+
+					if(!var31 && !var29) {
+						this.field_22375_o = this.field_22354_E;
+					} else {
+						this.field_22375_o = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4 + 1);
+					}
+
+					if(!var53 && !var28) {
+						this.field_22369_u = this.field_22358_C;
+					} else {
+						this.field_22369_u = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4 - 1);
+					}
+
+					if(!var31 && !var28) {
+						this.field_22367_w = this.field_22354_E;
+					} else {
+						this.field_22367_w = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4 + 1);
+					}
+
+					var55 = var1.getBlockBrightness(this.blockAccess, var47, var3, var4);
+					var12 = (this.field_22376_n + this.field_22375_o + var55 + this.field_22354_E) / 4.0F;
+					var9 = (var55 + this.field_22354_E + this.field_22368_v + this.field_22367_w) / 4.0F;
+					var10 = (this.field_22358_C + var55 + this.field_22369_u + this.field_22368_v) / 4.0F;
+					var11 = (this.field_22377_m + this.field_22376_n + this.field_22358_C + var55) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5 * 0.6F;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6 * 0.6F;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7 * 0.6F;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				var50 = var1.getBlockTextureFromSideAndMetadata(4, var21);
+				var51 = (var50 & 15) << 4;
+				var54 = var50 & 240;
+				var33 = (float)var2;
+				var34 = (float)var2;
+				var35 = (float)(var4 + 1);
+				var36 = (float)var4;
+				var19 = true;
+				var37 = (double)((float)(var51 + 0) / 256.0F);
+				var39 = ((double)(var51 + 16) - 0.01D) / 256.0D;
+				var41 = (double)(((float)var54 + (1.0F - var23) * 16.0F) / 256.0F);
+				var43 = (double)(((float)var54 + (1.0F - var22) * 16.0F) / 256.0F);
+				var45 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var8.setColorOpaque_F(this.colorRedTopLeft, this.colorGreenTopLeft, this.colorBlueTopLeft);
+				var8.addVertexWithUV((double)var33, (double)((float)var3 + var23), (double)var35, var37, var41);
+				var8.setColorOpaque_F(this.colorRedBottomLeft, this.colorGreenBottomLeft, this.colorBlueBottomLeft);
+				var8.addVertexWithUV((double)var34, (double)((float)var3 + var22), (double)var36, var39, var43);
+				var8.setColorOpaque_F(this.colorRedBottomRight, this.colorGreenBottomRight, this.colorBlueBottomRight);
+				var8.addVertexWithUV((double)var34, (double)(var3 + 0), (double)var36, var39, var45);
+				var8.setColorOpaque_F(this.colorRedTopRight, this.colorGreenTopRight, this.colorBlueTopRight);
+				var8.addVertexWithUV((double)var33, (double)(var3 + 0), (double)var35, var37, var45);
+			}
+
+			if(this.renderAllFaces || var18) {
+				var47 = var2 + 1;
+				if(this.field_22352_G <= 0) {
+					var12 = var1.getBlockBrightness(this.blockAccess, var47, var3, var4);
+					var11 = var12;
+					var10 = var12;
+					var9 = var12;
+				} else {
+					this.field_22371_s = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4);
+					this.field_22356_D = var1.getBlockBrightness(this.blockAccess, var47, var3, var4 - 1);
+					this.field_22353_F = var1.getBlockBrightness(this.blockAccess, var47, var3, var4 + 1);
+					this.field_22364_z = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4);
+					var28 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3 + 1, var4)];
+					var29 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3 - 1, var4)];
+					var53 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3, var4 + 1)];
+					var31 = Block.canBlockGrass[this.blockAccess.getBlockId(var47, var3, var4 - 1)];
+					if(!var29 && !var31) {
+						this.field_22372_r = this.field_22356_D;
+					} else {
+						this.field_22372_r = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4 - 1);
+					}
+
+					if(!var29 && !var53) {
+						this.field_22370_t = this.field_22353_F;
+					} else {
+						this.field_22370_t = var1.getBlockBrightness(this.blockAccess, var47, var3 - 1, var4 + 1);
+					}
+
+					if(!var28 && !var31) {
+						this.field_22365_y = this.field_22356_D;
+					} else {
+						this.field_22365_y = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4 - 1);
+					}
+
+					if(!var28 && !var53) {
+						this.field_22360_B = this.field_22353_F;
+					} else {
+						this.field_22360_B = var1.getBlockBrightness(this.blockAccess, var47, var3 + 1, var4 + 1);
+					}
+
+					var55 = var1.getBlockBrightness(this.blockAccess, var47, var3, var4);
+					var9 = (this.field_22371_s + this.field_22370_t + var55 + this.field_22353_F) / 4.0F;
+					var12 = (var55 + this.field_22353_F + this.field_22364_z + this.field_22360_B) / 4.0F;
+					var11 = (this.field_22356_D + var55 + this.field_22365_y + this.field_22364_z) / 4.0F;
+					var10 = (this.field_22372_r + this.field_22371_s + this.field_22356_D + var55) / 4.0F;
+				}
+
+				this.colorRedTopLeft = this.colorRedBottomLeft = this.colorRedBottomRight = this.colorRedTopRight = var5 * 0.6F;
+				this.colorGreenTopLeft = this.colorGreenBottomLeft = this.colorGreenBottomRight = this.colorGreenTopRight = var6 * 0.6F;
+				this.colorBlueTopLeft = this.colorBlueBottomLeft = this.colorBlueBottomRight = this.colorBlueTopRight = var7 * 0.6F;
+				this.colorRedTopLeft *= var9;
+				this.colorGreenTopLeft *= var9;
+				this.colorBlueTopLeft *= var9;
+				this.colorRedBottomLeft *= var10;
+				this.colorGreenBottomLeft *= var10;
+				this.colorBlueBottomLeft *= var10;
+				this.colorRedBottomRight *= var11;
+				this.colorGreenBottomRight *= var11;
+				this.colorBlueBottomRight *= var11;
+				this.colorRedTopRight *= var12;
+				this.colorGreenTopRight *= var12;
+				this.colorBlueTopRight *= var12;
+				var50 = var1.getBlockTextureFromSideAndMetadata(5, var21);
+				var51 = (var50 & 15) << 4;
+				var54 = var50 & 240;
+				var33 = (float)(var2 + 1);
+				var34 = (float)(var2 + 1);
+				var35 = (float)var4;
+				var36 = (float)(var4 + 1);
+				var19 = true;
+				var37 = (double)((float)(var51 + 0) / 256.0F);
+				var39 = ((double)(var51 + 16) - 0.01D) / 256.0D;
+				var41 = (double)(((float)var54 + (1.0F - var25) * 16.0F) / 256.0F);
+				var43 = (double)(((float)var54 + (1.0F - var24) * 16.0F) / 256.0F);
+				var45 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var8.setColorOpaque_F(this.colorRedBottomRight, this.colorGreenBottomRight, this.colorBlueBottomRight);
+				var8.addVertexWithUV((double)var33, (double)((float)var3 + var25), (double)var35, var37, var41);
+				var8.setColorOpaque_F(this.colorRedTopRight, this.colorGreenTopRight, this.colorBlueTopRight);
+				var8.addVertexWithUV((double)var34, (double)((float)var3 + var24), (double)var36, var39, var43);
+				var8.setColorOpaque_F(this.colorRedTopLeft, this.colorGreenTopLeft, this.colorBlueTopLeft);
+				var8.addVertexWithUV((double)var34, (double)(var3 + 0), (double)var36, var39, var45);
+				var8.setColorOpaque_F(this.colorRedBottomLeft, this.colorGreenBottomLeft, this.colorBlueBottomLeft);
+				var8.addVertexWithUV((double)var33, (double)(var3 + 0), (double)var35, var37, var45);
+			}
+
+			var1.minY = 0.0D;
+			var1.maxY = 1.0D;
+			return var19;
+		}
+	}
+
+	public boolean renderBlockFluidsWithColorMultiplier(Block var1, int var2, int var3, int var4, float var5, float var6, float var7) {
+		Tessellator var8 = Tessellator.instance;
+		boolean var9 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 + 1, var4, 1);
+		boolean var10 = var1.shouldSideBeRendered(this.blockAccess, var2, var3 - 1, var4, 0);
+		boolean var11 = var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 - 1, 2);
+		boolean var12 = var1.shouldSideBeRendered(this.blockAccess, var2, var3, var4 + 1, 3);
+		boolean var13 = var1.shouldSideBeRendered(this.blockAccess, var2 - 1, var3, var4, 4);
+		boolean var14 = var1.shouldSideBeRendered(this.blockAccess, var2 + 1, var3, var4, 5);
+		if(!this.renderAllFaces && !var9 && !var10 && !var11 && !var12 && !var13 && !var14) {
+			return false;
+		} else {
+			boolean var15 = false;
+			float var16 = 0.5F;
+			float var17 = 1.0F;
+			float var18 = 0.8F;
+			float var19 = 0.6F;
+			double var20 = 0.0D;
+			double var22 = 1.0D;
+			Material var24 = var1.blockMaterial;
+			int var25 = this.blockAccess.getBlockMetadata(var2, var3, var4);
+			float var26 = this.func_1224_a(var2, var3, var4, var24);
+			float var27 = this.func_1224_a(var2, var3, var4 + 1, var24);
+			float var28 = this.func_1224_a(var2 + 1, var3, var4 + 1, var24);
+			float var29 = this.func_1224_a(var2 + 1, var3, var4, var24);
+			int var30;
+			int var32;
+			int var33;
 			float var38;
+			float var39;
+			float var40;
+			if(this.renderAllFaces || var9) {
+				var15 = true;
+				var30 = var1.getBlockTextureFromSideAndMetadata(1, var25);
+				float var31 = (float)BlockFluid.func_293_a(this.blockAccess, var2, var3, var4, var24);
+				if(var31 > -999.0F) {
+					var30 = var1.getBlockTextureFromSideAndMetadata(2, var25);
+				}
+
+				var32 = (var30 & 15) << 4;
+				var33 = var30 & 240;
+				double var34;
+				double var36;
+				if(var31 < -999.0F) {
+					var34 = ((double)var32 + 8.0D) / 256.0D;
+					var36 = ((double)var33 + 8.0D) / 256.0D;
+					var38 = 0.0F;
+					var39 = 0.03125F;
+				} else {
+					var34 = (double)((float)(var32 + 16) / 256.0F);
+					var36 = (double)((float)(var33 + 16) / 256.0F);
+					var38 = MathHelper.sin(var31) * 8.0F / 256.0F;
+					var39 = MathHelper.cos(var31) * 8.0F / 256.0F;
+				}
+
+				var40 = var1.getBlockBrightness(this.blockAccess, var2, var3, var4);
+				var8.setColorOpaque_F(var17 * var40 * var5, var17 * var40 * var6, var17 * var40 * var7);
+				var8.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var26), (double)(var4 + 0), var34 - (double)var39 - (double)var38, var36 - (double)var39 + (double)var38);
+				var8.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var27), (double)(var4 + 1), var34 - (double)var39 + (double)var38, var36 + (double)var39 + (double)var38);
+				var8.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var28), (double)(var4 + 1), var34 + (double)var39 + (double)var38, var36 + (double)var39 - (double)var38);
+				var8.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var29), (double)(var4 + 0), var34 + (double)var39 - (double)var38, var36 - (double)var39 - (double)var38);
+			}
+
 			if(this.renderAllFaces || var10) {
-				var13 = true;
-				var28 = var1.getBlockTextureFromSideAndMetadata(1, var23);
-				float var29 = (float)BlockFluid.func_293_a(this.blockAccess, var2, var3, var4, var22);
-				if(var29 > -999.0F) {
-					var28 = var1.getBlockTextureFromSideAndMetadata(2, var23);
-				}
-
-				int var30 = (var28 & 15) << 4;
-				var31 = var28 & 240;
-				double var32 = ((double)var30 + 8.0D) / 256.0D;
-				double var34 = ((double)var31 + 8.0D) / 256.0D;
-				if(var29 < -999.0F) {
-					var29 = 0.0F;
-				} else {
-					var32 = (double)((float)(var30 + 16) / 256.0F);
-					var34 = (double)((float)(var31 + 16) / 256.0F);
-				}
-
-				var36 = MathHelper.sin(var29) * 8.0F / 256.0F;
-				var37 = MathHelper.cos(var29) * 8.0F / 256.0F;
-				var38 = var1.getBlockBrightness(this.blockAccess, var2, var3, var4);
-				var5.setColorOpaque_F(var15 * var38 * var7, var15 * var38 * var8, var15 * var38 * var9);
-				var5.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var24), (double)(var4 + 0), var32 - (double)var37 - (double)var36, var34 - (double)var37 + (double)var36);
-				var5.addVertexWithUV((double)(var2 + 0), (double)((float)var3 + var25), (double)(var4 + 1), var32 - (double)var37 + (double)var36, var34 + (double)var37 + (double)var36);
-				var5.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var26), (double)(var4 + 1), var32 + (double)var37 + (double)var36, var34 + (double)var37 - (double)var36);
-				var5.addVertexWithUV((double)(var2 + 1), (double)((float)var3 + var27), (double)(var4 + 0), var32 + (double)var37 - (double)var36, var34 - (double)var37 - (double)var36);
-			}
-
-			if(this.renderAllFaces || var11) {
 				float var52 = var1.getBlockBrightness(this.blockAccess, var2, var3 - 1, var4);
-				var5.setColorOpaque_F(var14 * var52, var14 * var52, var14 * var52);
+				var8.setColorOpaque_F(var16 * var52 * var5, var16 * var52 * var6, var16 * var52 * var7);
 				this.renderBottomFace(var1, (double)var2, (double)var3, (double)var4, var1.getBlockTextureFromSide(0));
-				var13 = true;
-			}
-
-			for(var28 = 0; var28 < 4; ++var28) {
-				int var53 = var2;
-				var31 = var4;
-				if(var28 == 0) {
-					var31 = var4 - 1;
-				}
-
-				if(var28 == 1) {
-					++var31;
-				}
-
-				if(var28 == 2) {
-					var53 = var2 - 1;
-				}
-
-				if(var28 == 3) {
-					++var53;
-				}
-
-				int var54 = var1.getBlockTextureFromSideAndMetadata(var28 + 2, var23);
-				int var33 = (var54 & 15) << 4;
-				int var55 = var54 & 240;
-				if(this.renderAllFaces || var12[var28]) {
-					float var35;
-					float var39;
-					float var40;
-					if(var28 == 0) {
-						var35 = var24;
-						var36 = var27;
-						var37 = (float)var2;
-						var39 = (float)(var2 + 1);
-						var38 = (float)var4;
-						var40 = (float)var4;
-					} else if(var28 == 1) {
-						var35 = var26;
-						var36 = var25;
-						var37 = (float)(var2 + 1);
-						var39 = (float)var2;
-						var38 = (float)(var4 + 1);
-						var40 = (float)(var4 + 1);
-					} else if(var28 == 2) {
-						var35 = var25;
-						var36 = var24;
-						var37 = (float)var2;
-						var39 = (float)var2;
-						var38 = (float)(var4 + 1);
-						var40 = (float)var4;
-					} else {
-						var35 = var27;
-						var36 = var26;
-						var37 = (float)(var2 + 1);
-						var39 = (float)(var2 + 1);
-						var38 = (float)var4;
-						var40 = (float)(var4 + 1);
-					}
-
-					var13 = true;
-					double var41 = (double)((float)(var33 + 0) / 256.0F);
-					double var43 = ((double)(var33 + 16) - 0.01D) / 256.0D;
-					double var45 = (double)(((float)var55 + (1.0F - var35) * 16.0F) / 256.0F);
-					double var47 = (double)(((float)var55 + (1.0F - var36) * 16.0F) / 256.0F);
-					double var49 = ((double)(var55 + 16) - 0.01D) / 256.0D;
-					float var51 = var1.getBlockBrightness(this.blockAccess, var53, var3, var31);
-					if(var28 < 2) {
-						var51 *= var16;
-					} else {
-						var51 *= var17;
-					}
-
-					var5.setColorOpaque_F(var15 * var51 * var7, var15 * var51 * var8, var15 * var51 * var9);
-					var5.addVertexWithUV((double)var37, (double)((float)var3 + var35), (double)var38, var41, var45);
-					var5.addVertexWithUV((double)var39, (double)((float)var3 + var36), (double)var40, var43, var47);
-					var5.addVertexWithUV((double)var39, (double)(var3 + 0), (double)var40, var43, var49);
-					var5.addVertexWithUV((double)var37, (double)(var3 + 0), (double)var38, var41, var49);
-				}
-			}
-
-			var1.minY = var18;
-			var1.maxY = var20;
-			return var13;
+				var15 = true;
+			}
+
+			float var37;
+			double var41;
+			double var43;
+			double var45;
+			double var47;
+			double var49;
+			float var51;
+			int var53;
+			int var54;
+			if(this.renderAllFaces || var11) {
+				var53 = var4 - 1;
+				var32 = var1.getBlockTextureFromSideAndMetadata(2, var25);
+				var33 = (var32 & 15) << 4;
+				var54 = var32 & 240;
+				var37 = (float)var2;
+				var38 = (float)(var2 + 1);
+				var39 = (float)var4;
+				var40 = (float)var4;
+				var15 = true;
+				var41 = (double)((float)(var33 + 0) / 256.0F);
+				var43 = ((double)(var33 + 16) - 0.01D) / 256.0D;
+				var45 = (double)(((float)var54 + (1.0F - var26) * 16.0F) / 256.0F);
+				var47 = (double)(((float)var54 + (1.0F - var29) * 16.0F) / 256.0F);
+				var49 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var51 = var1.getBlockBrightness(this.blockAccess, var2, var3, var53);
+				var51 *= var18;
+				var8.setColorOpaque_F(var17 * var51 * var5, var17 * var51 * var6, var17 * var51 * var7);
+				var8.addVertexWithUV((double)var37, (double)((float)var3 + var26), (double)var39, var41, var45);
+				var8.addVertexWithUV((double)var38, (double)((float)var3 + var29), (double)var40, var43, var47);
+				var8.addVertexWithUV((double)var38, (double)(var3 + 0), (double)var40, var43, var49);
+				var8.addVertexWithUV((double)var37, (double)(var3 + 0), (double)var39, var41, var49);
+			}
+
+			if(this.renderAllFaces || var12) {
+				var53 = var4 + 1;
+				var32 = var1.getBlockTextureFromSideAndMetadata(3, var25);
+				var33 = (var32 & 15) << 4;
+				var54 = var32 & 240;
+				var37 = (float)(var2 + 1);
+				var38 = (float)var2;
+				var39 = (float)(var4 + 1);
+				var40 = (float)(var4 + 1);
+				var15 = true;
+				var41 = (double)((float)(var33 + 0) / 256.0F);
+				var43 = ((double)(var33 + 16) - 0.01D) / 256.0D;
+				var45 = (double)(((float)var54 + (1.0F - var28) * 16.0F) / 256.0F);
+				var47 = (double)(((float)var54 + (1.0F - var27) * 16.0F) / 256.0F);
+				var49 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var51 = var1.getBlockBrightness(this.blockAccess, var2, var3, var53);
+				var51 *= var18;
+				var8.setColorOpaque_F(var17 * var51 * var5, var17 * var51 * var6, var17 * var51 * var7);
+				var8.addVertexWithUV((double)var37, (double)((float)var3 + var28), (double)var39, var41, var45);
+				var8.addVertexWithUV((double)var38, (double)((float)var3 + var27), (double)var40, var43, var47);
+				var8.addVertexWithUV((double)var38, (double)(var3 + 0), (double)var40, var43, var49);
+				var8.addVertexWithUV((double)var37, (double)(var3 + 0), (double)var39, var41, var49);
+			}
+
+			if(this.renderAllFaces || var13) {
+				var30 = var2 - 1;
+				var32 = var1.getBlockTextureFromSideAndMetadata(4, var25);
+				var33 = (var32 & 15) << 4;
+				var54 = var32 & 240;
+				var37 = (float)var2;
+				var38 = (float)var2;
+				var39 = (float)(var4 + 1);
+				var40 = (float)var4;
+				var15 = true;
+				var41 = (double)((float)(var33 + 0) / 256.0F);
+				var43 = ((double)(var33 + 16) - 0.01D) / 256.0D;
+				var45 = (double)(((float)var54 + (1.0F - var27) * 16.0F) / 256.0F);
+				var47 = (double)(((float)var54 + (1.0F - var26) * 16.0F) / 256.0F);
+				var49 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var51 = var1.getBlockBrightness(this.blockAccess, var30, var3, var4);
+				var51 *= var19;
+				var8.setColorOpaque_F(var17 * var51 * var5, var17 * var51 * var6, var17 * var51 * var7);
+				var8.addVertexWithUV((double)var37, (double)((float)var3 + var27), (double)var39, var41, var45);
+				var8.addVertexWithUV((double)var38, (double)((float)var3 + var26), (double)var40, var43, var47);
+				var8.addVertexWithUV((double)var38, (double)(var3 + 0), (double)var40, var43, var49);
+				var8.addVertexWithUV((double)var37, (double)(var3 + 0), (double)var39, var41, var49);
+			}
+
+			if(this.renderAllFaces || var14) {
+				var30 = var2 + 1;
+				var32 = var1.getBlockTextureFromSideAndMetadata(5, var25);
+				var33 = (var32 & 15) << 4;
+				var54 = var32 & 240;
+				var37 = (float)(var2 + 1);
+				var38 = (float)(var2 + 1);
+				var39 = (float)var4;
+				var40 = (float)(var4 + 1);
+				var15 = true;
+				var41 = (double)((float)(var33 + 0) / 256.0F);
+				var43 = ((double)(var33 + 16) - 0.01D) / 256.0D;
+				var45 = (double)(((float)var54 + (1.0F - var29) * 16.0F) / 256.0F);
+				var47 = (double)(((float)var54 + (1.0F - var28) * 16.0F) / 256.0F);
+				var49 = ((double)(var54 + 16) - 0.01D) / 256.0D;
+				var51 = var1.getBlockBrightness(this.blockAccess, var30, var3, var4);
+				var51 *= var19;
+				var8.setColorOpaque_F(var17 * var51 * var5, var17 * var51 * var6, var17 * var51 * var7);
+				var8.addVertexWithUV((double)var37, (double)((float)var3 + var29), (double)var39, var41, var45);
+				var8.addVertexWithUV((double)var38, (double)((float)var3 + var28), (double)var40, var43, var47);
+				var8.addVertexWithUV((double)var38, (double)(var3 + 0), (double)var40, var43, var49);
+				var8.addVertexWithUV((double)var37, (double)(var3 + 0), (double)var39, var41, var49);
+			}
+
+			var1.minY = var20;
+			var1.maxY = var22;
+			return var15;
 		}
 	}
 
@@ -3182,17 +3781,51 @@
 		int var5;
 		float var6;
 		float var7;
+		float var8;
 		if(this.field_31088_b) {
 			var5 = var1.getRenderColor(var2);
 			var6 = (float)(var5 >> 16 & 255) / 255.0F;
 			var7 = (float)(var5 >> 8 & 255) / 255.0F;
-			float var8 = (float)(var5 & 255) / 255.0F;
+			var8 = (float)(var5 & 255) / 255.0F;
 			GL11.glColor4f(var6 * var3, var7 * var3, var8 * var3, 1.0F);
 		}
 
 		var5 = var1.getRenderType();
 		if(var5 != 0 && var5 != 16) {
-			if(var5 == 1) {
+			if(var5 == 4) {
+				var8 = 0.01F;
+				float var9 = 0.1F;
+				var1.setBlockBounds(var8 + 0.0F, 0.0F - var9 + var8, var8 + 0.0F, var8 + 1.0F, 1.0F - var9 + var8, var8 + 1.0F);
+				GL11.glTranslatef(-0.5F, -0.5F, -0.5F);
+				var4.startDrawingQuads();
+				var4.setNormal(0.0F, -1.0F, 0.0F);
+				this.renderBottomFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(0));
+				var4.draw();
+				var4.startDrawingQuads();
+				var4.setNormal(0.0F, 1.0F, 0.0F);
+				this.renderTopFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(1));
+				var4.draw();
+				var4.startDrawingQuads();
+				var4.setNormal(0.0F, 0.0F, -1.0F);
+				this.renderEastFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(2));
+				var4.draw();
+				var4.startDrawingQuads();
+				var4.setNormal(0.0F, 0.0F, 1.0F);
+				this.renderWestFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(3));
+				var4.draw();
+				var4.startDrawingQuads();
+				var4.setNormal(-1.0F, 0.0F, 0.0F);
+				this.renderNorthFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(4));
+				var4.draw();
+				var4.startDrawingQuads();
+				var4.setNormal(1.0F, 0.0F, 0.0F);
+				this.renderSouthFace(var1, 0.0D, 0.0D, 0.0D, var1.getBlockTextureFromSide(5));
+				var4.draw();
+				GL11.glTranslatef(0.5F, 0.5F, 0.5F);
+				float var10 = 0.0F;
+				float var11 = 0.0F;
+				var1.setBlockBounds(0.0F + var11, 0.0F + var10, 0.0F + var11, 1.0F + var11, 1.0F + var10, 1.0F + var11);
+			} else if(var5 == 1) {
 				var4.startDrawingQuads();
 				var4.setNormal(0.0F, -1.0F, 0.0F);
 				this.renderCrossedSquares(var1, var2, -0.5D, -0.5D, -0.5D);
@@ -3371,6 +4004,6 @@
 	}
 
 	public static boolean renderItemIn3d(int var0) {
-		return var0 == 0 ? true : (var0 == 13 ? true : (var0 == 10 ? true : (var0 == 11 ? true : var0 == 16)));
+		return var0 == 0 ? true : (var0 == 4 ? true : (var0 == 13 ? true : (var0 == 10 ? true : (var0 == 11 ? true : var0 == 16))));
 	}
 }
--- net/minecraft/src/MaterialLiquid.java
+++ net/minecraft/src/MaterialLiquid.java
@@ -15,6 +15,10 @@
 		return false;
 	}
 
+	public boolean getCanBlockGrass() {
+		return false;
+	}
+
 	public boolean isSolid() {
 		return false;
 	}
